---
title: "GWAS Report"
output: html_notebook
params:
  data_path: DOV_DOG_BEN_APA_BIK_BAVA_BIR_TP1_BP2_WOB9
  output_file: NULL
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(tidyverse)
require(data.table)
require(knitr)
require(stringr)
require(kableExtra)
require(GenomicRanges)
require(rtracklayer)
require(gggenes)
require(rentrez)
knitr::opts_chunk$set(dev="png", 
                      dev.args=list(type="cairo"))
source(file.path(root,"code","06_Association","gemma_gwas_functions.R"))
```

```{r}
anno_file<-file.path(root,"input_data","00_Reference_Genome","MIC4273_HAP1_MSU618.liftoff.renamed.sorted.genes_only.gff")
feat_table_file<-file.path(root,"input_data","00_Reference_Genome","GCF_002872115.1_PKINGS_0.1_feature_table.txt")
gemma_file<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,".lmm2.assoc.txt"))
minpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_permution"),"min.txt")


annotations<-import(anno_file)
strand(annotations)<-"*"


feat_table<-fread(feat_table_file)
pk.lmm<- gemma.order(gemma_file, "p_lrt")
min_data<-read.csv(minpath,header=F,sep="\t")
pvals<-sort(min_data$V10, decreasing = F)
significant = -log10(pvals[length(pvals) - floor(length(pvals)*0.95)])
suggestive = -log10(pvals[length(pvals) - floor(length(pvals)*0.67)])
significant
suggestive
```

```{r}
#remove low log p to reduce memory footprint
pk.lmm.filt <- pk.lmm %>% filter(log_p > 1) %>% filter((!grepl("ups", chr)) & !is.na(row) & !is.na(log_p))

pk.lmm.filt<- pk.lmm.filt %>% mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

```

```{r, fig.width=20,fig.height=5}
p.gemma <- manc2.labels(pk.lmm.filt, "log_p")
p.gemma +
  geom_hline(yintercept = suggestive, linetype = 2, color = "blue") +
  geom_hline(yintercept = significant, linetype = 2, color = "red") +
    geom_point(data = subset(pk.lmm.filt, log_p > suggestive), aes(x = row, y = log_p,size=log_p), color = "red", shape=20,stroke=0.2,alpha=0.25) +
  scale_size_continuous(range = c(0.9, 3)) +
  labs(y = expression(paste("-log"[10], italic("P"),"-value")),x= "Chromosome")+
  ylim(c(1,15))
```


```{r}
ldpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_index_snps.ld.ld"))
ld_data<-fread(ldpath)
gwas_ld_data<-merge(ld_data,pk.lmm.filt,by.x="SNP_B",by.y="new_rs")
```

```{r}
peak_data_path<-file.path(root,"output_data","06_Association",params$data_path)
pk.lmm.peaks.out<-read.table(file.path(peak_data_path,paste0(params$data_path,"_SNPS_IN_PEAKS.txt")),header=T)

index_snps <- pk.lmm.peaks.out %>%
  group_by(peak) %>%
  # Filter for rows with the maximum log_p value within each peak
  filter(log_p == max(log_p)) %>%
  # Sample one row randomly in case of ties
  slice_sample(n = 1) %>%
  ungroup() %>% # Remove the grouping structure
  mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

index_snps_simp<-index_snps[,c(16,17,1,3,10)]
colnames(index_snps_simp)<-c("PEAK","SNP_A","CHR_A","BP_A","p_lrt")
index_snps_simp$maximum<-index_snps_simp$BP_A+75000
index_snps_simp$minimum<-index_snps_simp$BP_A-75000

gwas_ld_data<-merge(gwas_ld_data,index_snps_simp[,1:2],by.x="SNP_A",by.y="SNP_A")

snp_gr <- GRanges(
  seqnames = index_snps_simp$CHR_A,
  ranges = IRanges(start = index_snps_simp$minimum, end = index_snps_simp$maximum),
)

mcols(snp_gr)$PEAK <- index_snps_simp$PEAK

# Perform the overlap
overlaps <- findOverlaps(snp_gr, annotations, ignore.strand = TRUE)

# Create a mapping table from overlaps
overlap_mapping <- data.frame(
  peak_index = queryHits(overlaps),
  annotation_index = subjectHits(overlaps)
)

overlap_mapping$PEAK <- mcols(snp_gr)$PEAK[overlap_mapping$peak_index]

# Convert annotations to a dataframe
overlapping_annotations_df <- as.data.frame(annotations)

# Ensure there's a unique identifier for each annotation
overlapping_annotations_df$unique_id <- seq_len(nrow(overlapping_annotations_df))

# Filter overlap_mapping to include only unique annotation indices
overlap_mapping_unique <- overlap_mapping[!duplicated(overlap_mapping$annotation_index), ]

# Merge SNP_A into overlapping_annotations_df
overlapping_annotations_df <- merge(overlapping_annotations_df, overlap_mapping_unique, by.x = "unique_id", by.y = "annotation_index", all.x = TRUE)
overlapping_annotations_df<-subset(overlapping_annotations_df,!is.na(PEAK))
```

```{r}
filtered_snps_list <- list()

# Loop through each row in index_snps_simp
for(i in 1:nrow(index_snps_simp)) {
  # Extract the current peak's details
  current_peak <- index_snps_simp[i, ]
  
  # Filter pk.lmm.filt for SNPs within the current peak's range
  filtered_snps <- pk.lmm %>%
    filter(chr == current_peak$CHR_A, ps >= current_peak$minimum, ps <= current_peak$maximum) %>%
    mutate(PEAK = current_peak$PEAK) # Add the peak identifier
  
  # Store the filtered data frame in the list
  filtered_snps_list[[i]] <- filtered_snps
}

# Combine all filtered data frames into one
all_filtered_snps <- do.call(rbind, filtered_snps_list)

# Optionally, convert the list back to a data frame or tibble if needed
all_filtered_snps_df <- as.data.frame(all_filtered_snps)
```

```{r, fig.width=12, fig.height=24}

filtered_gwas_ld_data <- gwas_ld_data %>%
  filter(abs(BP_B - BP_A) <= 75000)


p<-ggplot()+
  geom_point(data=all_filtered_snps_df, mapping=aes(x=ps,y=-log10(p_lrt)),color="grey")+
  geom_point(data=filtered_gwas_ld_data , mapping=aes(x=BP_B, y=-log10(p_lrt),color=R2)) +
  scale_color_gradientn(colours = c("blue","green","yellow","red")) +
  geom_hline(yintercept=significant, linetype="dashed", color="red", size=0.5) +
  geom_hline(yintercept=suggestive, linetype="dashed", color="blue", size=0.5) +
  geom_point(data=index_snps_simp,
             mapping=aes(x=BP_A,y=-log10(p_lrt)),shape=5,size=5,fill="black") +
  geom_gene_arrow(data=overlapping_annotations_df, mapping=aes(xmin=start,xmax=end,y=-1,fill=gene),show.legend =FALSE) +
  geom_gene_label(data=overlapping_annotations_df, mapping=aes(xmin=start,xmax=end,y=-1,label=gene),align="left",show.legend = FALSE)+
  scale_x_continuous(labels = function(l) {trans = l /1e6; paste0(trans, "mb")})
  #scale_fill_brewer(palette = "Set3") +
  #theme_genes()

p+facet_wrap(~PEAK, scales="free_x") + theme(axis.text.x = element_text(angle = 90, hjust = 1))

```

## Top Scoring SNPs and Nearest Genes:
```{r, echo = FALSE, results='asis'}
overlapping_annotations_df$gid<-as.numeric(str_split(unlist(overlapping_annotations_df$Dbxref),":",simplify=TRUE)[,2])

gene_summaries <- entrez_summary(db = "gene", id = overlapping_annotations_df$gid)
gene_info <- lapply(gene_summaries, function(x) {
  c(GeneID = x$uid, Symbol = x$name, Description = x$title, description=x$description)
})

# Convert list to data frame for easier viewing/manipulation
gene_info_df <- do.call(rbind, gene_info)
rownames(gene_info_df) <- NULL
gene_info_df<-as.data.frame(gene_info_df)
gene_info_df$GeneID<-as.numeric(gene_info_df$GeneID)


suggestive_snps<-NULL
snp_count_per_peak<-NULL
genes_per_peak<-NULL
final_report<-NULL
kable_table<-NULL

suggestive_snps <- gwas_ld_data %>%
  filter(-log10(p_lrt) > suggestive)

snps_count_per_peak <- suggestive_snps %>%
  inner_join(index_snps_simp, by = c("PEAK" = "PEAK")) %>%
  group_by(PEAK) %>%
  summarise(SNP_Count = n())

genes_per_peak <- overlapping_annotations_df %>%
  left_join(gene_info_df, by = c("gid" = "GeneID")) %>%
  group_by(PEAK) %>%
  summarise(Associated_Genes = toString(Symbol), Names= toString(description))

final_report <- index_snps_simp %>%
  select(PEAK, CHR_A, maximum, minimum) %>%
  left_join(snps_count_per_peak, by = "PEAK") %>%
  left_join(genes_per_peak, by = "PEAK")

colnames(final_report)<-c("Peak", "Chromosome","Start","End","n SNPs","Associated Genes", "Descriptions")
# Create the kable table

kable_table <- kable(final_report, escape = FALSE) %>%
  kable_styling(bootstrap_options = c("striped"), full_width = T) %>%
  column_spec(1, bold = T) %>% # Example to bold the first column, adjust as needed
  row_spec(0, background = "#D3D3D3") # Example to change header row background

# Print the table to view in R Markdown or an HTML document
kable_table
```
