---
title: "Peak Report"
output: html_notebook
params:
  data_path: MOV_BIR_DOG_DOV_BAVA_BIK_APA_BEN_TP1_BP2_WOB9
  output_file: NULL
  popgen_prefix: MIC4273_HAP2_NEWREF_MSU618_all_fish_all_sites
  MAF_thresh: 0.05
  pops: ["MOV","BIR","DOG","DOV","BAVA","BIK","APA","BEN"]
  phenos: ["TP","BP","TP","TP","BP","BP","TP","TP"]
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(tidyverse)
library(patchwork)
library(data.table)
library(reshape2)
library(stringr)
library(scales)      # For percent_format in scale_y_continuous
require(GenomicRanges)
require(rtracklayer)
require(gggenes)
require(rentrez)
require(rlang)
require(biomaRt)
library(BioMartGOGeneSets)

source(file.path(root,"code","06_Association","gemma_gwas_functions.R"))
source(file.path(root,"code","08_Peak_Analysis","peak_analysis_functions.R"))

```

## Data Loading

### Load the Genotypic Data
```{r}
geno_path<-file.path(root,"output_data","08_Peak_Analysis",params$data_path,paste0(params$data_path,".bin.pos.genos"))
peaks_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_PEAKS.bed"))
logp_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_SNPS_IN_PEAKS_TOP250.txt"))
dxy_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_dxy.txt"))
fst_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_fst.txt"))
pi_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_pi.txt"))
tajd_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_tajd.txt"))


df.genos<-read.table(geno_path,header=T)

long_genotype <- df.genos %>%
  pivot_longer(
    cols = -c(CHROM, POS, REF, ALT), # Exclude the first four columns
    names_to = "name",          # Column for the new 'name' variable
    values_to = "geno"          # Column for the new 'geno' variable
  ) %>%
  dplyr::select(name, chr = CHROM, pos = POS, geno) # Rename and select columns

long_genotype<-as.data.frame(long_genotype)

long_genotype<-long_genotype %>% mutate(chr=str_replace(chr,"chr",""))

df.bed <- read.table(peaks_path)
names(df.bed) <- c("chr", "start","end", "peak")

#remove positions if there is only one SNP
df.bed <- df.bed %>% 
  mutate(lg = end- start) %>% 
  filter(lg > 0)

#load the top p values (filtered > 0.1) to plot the signal with the other plots I generate below
df.logp <- read.table(logp_path)

```

### Load the Population Genetics Statistics
```{r}
pixy_df <- list()

tajd_df<-fread(tajd_file)

tajd_df<-tajd_df %>% arrange(pop1,CHROM,BIN_START)

tajd_df <- tajd_df %>%
    mutate(window_pos_1=BIN_START+1) %>%
    mutate(window_pos_2=BIN_START+5000) %>%
    mutate(chromosome=CHROM) %>%
    mutate(statistic="avg_tajd") %>%
    mutate(value=TajimaD) %>%
    mutate(pop2=NA) %>%
    dplyr::select(pop1,  pop2, chromosome, window_pos_1, window_pos_2, statistic, value, ) %>%
    arrange(pop1,  pop2, chromosome, window_pos_1)

df_pi <- fread(pi_file)
      
df_pi <- df_pi %>%
  gather(-pop, -window_pos_1, -window_pos_2, -chromosome,
  key = "statistic", value = "value") %>%
  mutate(pop1= pop) %>%
  mutate(pop2 = NA) %>%
  dplyr::select(pop1, chromosome, window_pos_1, window_pos_2, statistic, value, pop2 )

fst_df <- fread(fst_file)

fst_df <- fst_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
dxy_df <- fread(dxy_file)

dxy_df <- dxy_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
      
pixy_df[[1]] <- fst_df
pixy_df[[2]] <- dxy_df
pixy_df[[3]] <- df_pi
pixy_df[[4]] <- tajd_df

pixy<-bind_rows(pixy_df) %>%
    arrange(pop1, pop2, chromosome, window_pos_1, statistic)


pixy<-pixy %>% pivot_wider(names_from=statistic,values_from=value) %>% mutate(comparison=paste0(pop1,"_",pop2))

pixy<-pixy %>% group_by(comparison) %>% mutate(zfst=((avg_wc_fst-mean(avg_wc_fst,na.rm=T))/sd(avg_wc_fst,na.rm=T)))

pixy <- pixy %>% 
  pivot_longer(cols = -c(pop1, pop2, chromosome, window_pos_1, window_pos_2,comparison),
               names_to = "statistic", values_to = "value")

rm(df_pi,dxy_df,fst_df,pixy_df,tajd_df)
gc()

pixy <- pixy %>%
  mutate(region_status = case_when(
    chromosome %in% df.bed$chr & window_pos_1 >= df.bed$start[match(chromosome, df.bed$chr)] & 
      window_pos_2 <= df.bed$end[match(chromosome, df.bed$chr)] ~ "Inside",
    TRUE ~ "Outside"
  ))
```

### Filter Population Genetics Statistics
```{r}
phenotypes<-data.frame(POP=params$pops,population_phenotype=params$phenos)

# Dynamically create the filter condition
filter_expr <- paste0("pop1 == '", params$pops, "'", collapse = " | ")

# Generate all pairwise combinations
pairs <- combn(params$pops, 2, simplify = FALSE)

# Create forward comparisons
forward <- sapply(pairs, function(x) paste(x, collapse = "_"))

# Create backward comparisons
backward <- sapply(pairs, function(x) paste(rev(x), collapse = "_"))

# Combine forward and backward comparisons
comparison_list <- c(forward, backward)

comparision_expr <- paste0("comparison == '", comparison_list, "'", collapse = " | ")

pi_plot <- pixy %>%
  filter(statistic == "avg_pi", !is.na(value)) %>%
  filter(!!rlang::parse_expr(filter_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(phenotype_pop1 = population_phenotype)

tajd_plot<-pixy %>%
    filter(statistic == "avg_tajd", !is.na(value)) %>%
    filter(!!rlang::parse_expr(filter_expr)) %>%
    mutate(value = as.numeric(value) ) %>%  
    left_join(phenotypes, by = c("pop1" = "POP")) %>%
    dplyr::rename(phenotype_pop1 = population_phenotype)

fst_plot <- pixy %>%
  filter(statistic == "avg_wc_fst", !is.na(value)) %>%
  filter(!!rlang::parse_expr(comparision_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(pop1_phenotype = population_phenotype) %>%
  left_join(phenotypes, by = c("pop2" = "POP")) %>%
  dplyr::rename(pop2_phenotype = population_phenotype) %>%
  mutate(
    comparison_classification = case_when(
      pop1_phenotype == pop2_phenotype ~ "Same phenotype",
      pop1_phenotype != pop2_phenotype ~ "Different phenotypes",
      TRUE ~ "Unknown"
    )) %>%
  mutate(
    compare_color = case_when(
      comparison_classification == "Same phenotype" & pop1_phenotype=="BP" ~ "BP-BP",
      comparison_classification == "Same phenotype" & pop1_phenotype=="TP" ~ "TP-TP",
      comparison_classification == "Different phenotypes" ~ "BP-TP"
    )
  )
  

```

### Load SNP Data
```{r}
anno_file<-file.path(root,"input_data","00_Reference_Genome","jordan-mic4273-mb-hirise-20xx3__06-13-2024__final_assembly.ensembl.liftoff.renamed.genes_only.gff")
gemma_file<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,".lmm2.assoc.txt"))
minpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_permution"),"min.txt")
ldpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_index_snps.ld.ld"))
peak_data_path<-file.path(root,"output_data","06_Association",params$data_path)

annotations<-import(anno_file)
strand(annotations)<-"*"

pk.lmm<- gemma.order(gemma_file, "p_lrt")
min_data<-read.csv(minpath,header=F,sep="\t")
pvals<-sort(min_data$V10, decreasing = F)
significant = -log10(pvals[length(pvals) - floor(length(pvals)*0.95)])
suggestive = -log10(pvals[length(pvals) - floor(length(pvals)*0.67)])


#remove low log p to reduce memory footprint
pk.lmm.filt <- pk.lmm %>% filter(log_p > 1) %>% filter(af > params$MAF_thresh) %>% filter((!grepl("ups", chr)) & !is.na(row) & !is.na(log_p)) %>% arrange(numeric_chr)

pk.lmm.filt<- pk.lmm.filt %>% mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

ld_data<-fread(ldpath)
gwas_ld_data<-merge(ld_data,pk.lmm.filt,by.x="SNP_B",by.y="new_rs")


pk.lmm.peaks.out<-read.table(file.path(peak_data_path,paste0(params$data_path,"_SNPS_IN_PEAKS.txt")),header=T)

index_snps <- pk.lmm.peaks.out %>%
  group_by(peak) %>%
  # Filter for rows with the maximum log_p value within each peak
  filter(log_p == max(log_p)) %>%
  # Sample one row randomly in case of ties
  slice_sample(n = 1) %>%
  ungroup() %>% # Remove the grouping structure
  mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

index_snps_simp<-index_snps[,c(16,17,1,3,10)]
colnames(index_snps_simp)<-c("PEAK","SNP_A","CHR_A","BP_A","p_lrt")
index_snps_simp$maximum<-index_snps_simp$BP_A+75000
index_snps_simp$minimum<-index_snps_simp$BP_A-75000

gwas_ld_data<-merge(gwas_ld_data,index_snps_simp[,1:2],by.x="SNP_A",by.y="SNP_A")

snp_gr <- GRanges(
  seqnames = index_snps_simp$CHR_A,
  ranges = IRanges(start = index_snps_simp$minimum, end = index_snps_simp$maximum),
)

mcols(snp_gr)$PEAK <- index_snps_simp$PEAK

# Perform the overlap
overlaps <- findOverlaps(snp_gr, annotations, ignore.strand = TRUE)

# Create a mapping table from overlaps
overlap_mapping <- data.frame(
  peak_index = queryHits(overlaps),
  annotation_index = subjectHits(overlaps)
)

overlap_mapping$PEAK <- mcols(snp_gr)$PEAK[overlap_mapping$peak_index]

# Convert annotations to a dataframe
overlapping_annotations_df <- as.data.frame(annotations)

# Ensure there's a unique identifier for each annotation
overlapping_annotations_df$unique_id <- seq_len(nrow(overlapping_annotations_df))

# Filter overlap_mapping to include only unique annotation indices
overlap_mapping_unique <- overlap_mapping[!duplicated(overlap_mapping$annotation_index), ]

# Merge SNP_A into overlapping_annotations_df
overlapping_annotations_df <- merge(overlapping_annotations_df, overlap_mapping_unique, by.x = "unique_id", by.y = "annotation_index", all.x = TRUE)
overlapping_annotations_df<-subset(overlapping_annotations_df,!is.na(PEAK))

gr=getBioMartGenes("pkingsleyae_gene_ensembl")

anno_meta<-as.data.frame(gr)[6:9]

overlapping_annotations_df<-left_join(overlapping_annotations_df,anno_meta,by=c("gene_id"="ensembl_gene_id"))
```


```{r}
fish_data_file<-file.path(root,"input_data/01_Terra/data_model/fish_data_2024a.txt")

fish_data<-read.csv(fish_data_file,row.names=1,sep="\t")

pheno_discrete <- fish_data %>% 
  mutate(Phenotype = factor(Phenotype)) %>%
  dplyr::select(Phenotype)

phenos<-setNames(as.character(pheno_discrete$Phenotype), row.names(pheno_discrete))
phenos <- gsub("TYPE_II", "TP", phenos)
phenos <- gsub("TYPE_I", "TP", phenos)
phenos<-as.factor(phenos)
## set colors for plotting 
col.pheno<-setNames(c("blue","red","purple"),levels(phenos)) 

phenos_df <- data.frame(Sample = names(phenos), Phenotype = as.factor(phenos))
```

## Plotting

### Plot Genotype Matrix
```{r}
for (peak.idx in unique(df.bed$peak)) {
  
  # Take bed file, pull out peak of interest then filter by it
  df.bed.idx <- df.bed %>% filter(peak == peak.idx)
  chr.idx <- unique(df.bed.idx$chr)
  print(paste0("running ", peak.idx, " on ", chr.idx))
  
  # Subset geno df by the peak region 
  df.genos.sub <- long_genotype %>% 
    filter(chr == chr.idx & pos > min(df.bed.idx$start) & pos < max(df.bed.idx$end)) %>%
    arrange(name, pos) %>%  # Sort by name and then by position
    group_by(name) %>%
    mutate(snp_order_within_individual = row_number()) %>%  # Create sequence number
    ungroup()
  
  df.genos.sub$geno <- factor(ifelse(is.na(df.genos.sub$geno), "NA", as.character(df.genos.sub$geno)),
                              levels = c("NA", "0", "1", "2"))
  
  # Step 1: Merge phenotype information
  df.genos.sub <- df.genos.sub %>%
    mutate(population = sub("_.*", "", name)) %>%  # Extract population from the name
    left_join(phenos_df, by = c("name" = "Sample"))  # Merge phenos_df to include Phenotype
  
  # Step 2: Sort by phenotype within each population
  df.genos.sub <- df.genos.sub %>%
    group_by(population) %>%
    arrange(Phenotype, name, .by_group = TRUE) %>%  # Sort by Phenotype and then name within each population
    ungroup() %>%
    mutate(name = factor(name, levels = rev(unique(name))))  # Reverse levels for top-to-bottom plotting
  
  # Step 3: Identify the first individual (top-to-bottom) for each population
  first_individuals <- df.genos.sub %>%
    group_by(population) %>%
    slice_head(n = 1) %>%  # Select the first individual per population
    dplyr::select(name, population)  # Keep only the name and population
  
  # Create a mapping for y-axis labels: Label only the first individual per population
  label_mapping <- setNames(first_individuals$population, first_individuals$name)
  
  # Step 4: Create the ggplot object
  p <- ggplot() +
    # Plot the phenotype column
    geom_tile(data = df.genos.sub,
              aes(x = -1, y = name, fill = Phenotype),  # Use a separate column for phenotype
              width = 1, height = 1) +
    # Plot the genotypes
    geom_tile(data = df.genos.sub,
              aes(x = snp_order_within_individual, y = name, fill = geno)) +
    # Set fill scales for genotype and phenotype
    scale_fill_manual(name = "Genotype/Phenotype",
                      values = c(col.pheno, "NA" = "white", "0" = "grey", "1" = "#1a80bb", "2" = "#f2c45f")) +
    labs(title = paste0("Peak ", peak.idx, " on Chr ", chr.idx),
         x = "SNP Order within Individual", 
         y = "Individual",
         fill = "Genotype") +
    scale_y_discrete(labels = function(names) {
      # Only label the first individual per population
      ifelse(names %in% names(label_mapping), label_mapping[names], "")
    }) +
    scale_x_continuous(expand = c(0, 0), limits = c(-2, max(df.genos.sub$snp_order_within_individual) + 1)) +
    theme_minimal() +
    theme(panel.grid = element_blank())
  
  # Step 5: Save the plot
  ggsave(file.path(root, "output_data", "08_Peak_Analysis", params$data_path, 
                   paste0(params$data_path, "_PEAK", peak.idx, "_CHR", chr.idx, ".png")), 
         plot = p, width = 8.5, height = 17)
}


```

```{r}
# Main loop
for (peak.idx in unique(df.bed$peak)) {
  df.bed.idx <- df.bed %>% filter(peak == peak.idx)
  chr.idx <- paste0("chr",unique(df.bed.idx$chr))
  start <- min(df.bed.idx$start)
  end <- max(df.bed.idx$end)
  
  pi_peak <- filter_data(pi_plot, chr.idx, start,end,window_size = 100000)
  tpp_plot <- filter_data(tajd_plot, chr.idx, start,end,window_size = 100000)
  fpp_plot <- filter_data(fst_plot, chr.idx, start,end,window_size = 100000)
  
  idx_snp_peak<-index_snps_simp %>% filter(CHR_A == chr.idx & BP_A >= min(df.bed.idx$start)-100000 & BP_A <= min(df.bed.idx$start)+100000)
  
  # Create each plot
  spp <- create_gwas_plot(pk.lmm, gwas_ld_data, idx_snp_peak, overlapping_annotations_df, chr.idx, start,end,window_size = 100000)
  ppp <- create_pi_plot(pi_peak, df.bed.idx, chr.idx)
  tpp <- create_tajima_plot(tpp_plot, df.bed.idx, chr.idx)
  fpp <- create_fst_plot(fpp_plot, df.bed.idx, chr.idx)
  
  combined_peak_plot <- spp / ppp / tpp / fpp +
    plot_annotation(paste0("Peak ", peak.idx), tag_levels = 'A') &
    theme(plot.tag = element_text(size = 48, face = "bold"))
  
  # Adjust relative heights of the plots
  combined_peak_plot <- combined_peak_plot + plot_layout(heights = c(2, 2, 2, 2))
  
  ggsave(
    file.path(root, "output_data", "08_Peak_Analysis", params$data_path,
              paste0(params$data_path, "_PEAK", peak.idx, "_CHR", chr.idx, "_peak_stats_plot.png")),
    plot = combined_peak_plot, width = 8.5, height = 17
  )
  }
```
