---
title: "Peak Report"
output: html_notebook
params:
  data_path: APA_BEN_BAM_TP1_BP2_WOB9
  output_file: NULL
  popgen_prefix: MIC4273_HAP2_NEWREF_MSU618_all_fish_all_sites
  MAF_thresh: 0.05
  pops: ["APA","BAM","BEN"]
  phenos: ["TP","BP","TP"]
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(tidyverse)
library(patchwork)
library(data.table)
library(reshape2)
library(stringr)
library(scales)      # For percent_format in scale_y_continuous
require(GenomicRanges)
require(rtracklayer)
require(gggenes)
require(rentrez)
require(rlang)
require(biomaRt)
library(BioMartGOGeneSets)
library(ggtree)
library(ape)
library(ggsignif)
library(fuzzyjoin)
library(rstatix)
library(ggpubr)

source(file.path(root,"code","06_Association","gemma_gwas_functions.R"))
source(file.path(root,"code","08_Peak_Analysis","peak_analysis_functions.R"))
source(file.path(root,"code","08_Peak_Analysis","phylogenoplot_multi.R"))

rename.tips <- function(phy, old_names, new_names) {
   mpos <- match(old_names,phy$tip.labels)
   phy$tip.labels[mpos] <- new_names
   return(phy)
}

```

## Data Loading

### Load the Genotypic Data
```{r}
geno_path<-file.path(root,"output_data","08_Peak_Analysis",params$data_path,paste0(params$data_path,".bin.pos.genos"))
peaks_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_PEAKS.bed"))
logp_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_SNPS_IN_PEAKS_TOP250.txt"))
dxy_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_dxy.txt"))
fst_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_fst.txt"))
pi_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_pi.txt"))
tajd_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_tajd.txt"))


df.genos<-read.table(geno_path,header=T)

long_genotype <- df.genos %>%
  pivot_longer(
    cols = -c(CHROM, POS, REF, ALT), # Exclude the first four columns
    names_to = "name",          # Column for the new 'name' variable
    values_to = "geno"          # Column for the new 'geno' variable
  ) %>%
  dplyr::select(name, chr = CHROM, pos = POS, geno) # Rename and select columns

long_genotype<-as.data.frame(long_genotype)

long_genotype<-long_genotype %>% mutate(chr=str_replace(chr,"chr",""))

df.bed <- read.table(peaks_path)
names(df.bed) <- c("chr", "start","end", "peak")

#remove positions if there is only one SNP
df.bed <- df.bed %>% 
  mutate(lg = end- start) %>% 
  filter(lg > 0)

#load the top p values (filtered > 0.1) to plot the signal with the other plots I generate below
df.logp <- read.table(logp_path)

```

### Load the Population Genetics Statistics
```{r}

pixy_df <- list()

tajd_df<-fread(tajd_file)

tajd_df<-tajd_df %>% arrange(pop1,CHROM,BIN_START)

tajd_df <- tajd_df %>%
    mutate(window_pos_1=BIN_START+1) %>%
    mutate(window_pos_2=BIN_START+5000) %>%
    mutate(chromosome=CHROM) %>%
    mutate(statistic="avg_tajd") %>%
    mutate(value=TajimaD) %>%
    mutate(pop2=NA) %>%
    dplyr::select(pop1,  pop2, chromosome, window_pos_1, window_pos_2, statistic, value, ) %>%
    arrange(pop1,  pop2, chromosome, window_pos_1)

df_pi <- fread(pi_file)
      
df_pi <- df_pi %>%
  gather(-pop, -window_pos_1, -window_pos_2, -chromosome,
  key = "statistic", value = "value") %>%
  mutate(pop1= pop) %>%
  mutate(pop2 = NA) %>%
  dplyr::select(pop1, chromosome, window_pos_1, window_pos_2, statistic, value, pop2 )

fst_df <- fread(fst_file)

fst_df <- fst_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
dxy_df <- fread(dxy_file)

dxy_df <- dxy_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
      
pixy_df[[1]] <- fst_df
pixy_df[[2]] <- dxy_df
pixy_df[[3]] <- df_pi
pixy_df[[4]] <- tajd_df

pixy<-bind_rows(pixy_df) %>%
    arrange(pop1, pop2, chromosome, window_pos_1, statistic)


pixy<-pixy %>% pivot_wider(names_from=statistic,values_from=value) %>% mutate(comparison=paste0(pop1,"_",pop2))

pixy <- pixy %>% 
  pivot_longer(cols = -c(pop1, pop2, chromosome, window_pos_1, window_pos_2,comparison),
               names_to = "statistic", values_to = "value")

rm(df_pi,dxy_df,fst_df,pixy_df,tajd_df)
gc()

```

### Filter Population Genetics Statistics
```{r}
phenotypes<-data.frame(POP=params$pops,population_phenotype=params$phenos)

# Dynamically create the filter condition
filter_expr <- paste0("pop1 == '", params$pops, "'", collapse = " | ")

# Generate all pairwise combinations
pairs <- combn(params$pops, 2, simplify = FALSE)

# Create forward comparisons
forward <- sapply(pairs, function(x) paste(x, collapse = "_"))

# Create backward comparisons
backward <- sapply(pairs, function(x) paste(rev(x), collapse = "_"))

# Combine forward and backward comparisons
comparison_list <- c(forward, backward)

comparision_expr <- paste0("comparison == '", comparison_list, "'", collapse = " | ")

pi_plot <- pixy %>%
  filter(statistic == "avg_pi", !is.na(value)) %>%
  filter(!!rlang::parse_expr(filter_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(phenotype_pop1 = population_phenotype)

tajd_plot<-pixy %>%
    filter(statistic == "avg_tajd", !is.na(value)) %>%
    filter(!!rlang::parse_expr(filter_expr)) %>%
    mutate(value = as.numeric(value) ) %>%  
    left_join(phenotypes, by = c("pop1" = "POP")) %>%
    dplyr::rename(phenotype_pop1 = population_phenotype)

fst_plot <- pixy %>%
  filter(statistic == "avg_wc_fst", !is.na(value)) %>%
  filter(!!rlang::parse_expr(comparision_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(pop1_phenotype = population_phenotype) %>%
  left_join(phenotypes, by = c("pop2" = "POP")) %>%
  dplyr::rename(pop2_phenotype = population_phenotype) %>%
  mutate(
    comparison_classification = case_when(
      pop1_phenotype == pop2_phenotype ~ "Same phenotype",
      pop1_phenotype != pop2_phenotype ~ "Different phenotypes",
      TRUE ~ "Unknown"
    )) %>%
  mutate(
    compare_color = case_when(
      comparison_classification == "Same phenotype" & pop1_phenotype=="BP" ~ "BP-BP",
      comparison_classification == "Same phenotype" & pop1_phenotype=="TP" ~ "TP-TP",
      comparison_classification == "Different phenotypes" ~ "BP-TP"
    )
  )
  

```

### Load SNP Data
```{r}
anno_file<-file.path(root,"input_data","00_Reference_Genome","jordan-mic4273-mb-hirise-20xx3__06-13-2024__final_assembly.ensembl.liftoff.renamed.genes_only.gff")
gemma_file<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,".lmm2.assoc.txt"))
minpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_permution"),"min.txt")
ldpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_index_snps.ld.ld"))
peak_data_path<-file.path(root,"output_data","06_Association",params$data_path)

annotations<-import(anno_file)
strand(annotations)<-"*"

pk.lmm<- gemma.order(gemma_file, "p_lrt")
min_data<-read.csv(minpath,header=F,sep="\t")
pvals<-sort(min_data$V10, decreasing = F)
significant = -log10(pvals[length(pvals) - floor(length(pvals)*0.95)])
suggestive = -log10(pvals[length(pvals) - floor(length(pvals)*0.67)])


#remove low log p to reduce memory footprint
pk.lmm.filt <- pk.lmm %>% filter(log_p > 1) %>% filter(af > params$MAF_thresh) %>% filter((!grepl("ups", chr)) & !is.na(row) & !is.na(log_p)) %>% arrange(numeric_chr)

pk.lmm.filt<- pk.lmm.filt %>% mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

ld_data<-fread(ldpath)
gwas_ld_data<-merge(ld_data,pk.lmm.filt,by.x="SNP_B",by.y="new_rs")


pk.lmm.peaks.out<-read.table(file.path(peak_data_path,paste0(params$data_path,"_SNPS_IN_PEAKS.txt")),header=T)

index_snps <- pk.lmm.peaks.out %>%
  group_by(peak) %>%
  # Filter for rows with the maximum log_p value within each peak
  filter(log_p == max(log_p)) %>%
  # Sample one row randomly in case of ties
  slice_sample(n = 1) %>%
  ungroup() %>% # Remove the grouping structure
  mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

index_snps_simp<-index_snps[,c(16,17,1,3,10)]
colnames(index_snps_simp)<-c("PEAK","SNP_A","CHR_A","BP_A","p_lrt")
index_snps_simp$maximum<-index_snps_simp$BP_A+75000
index_snps_simp$minimum<-index_snps_simp$BP_A-75000

gwas_ld_data<-merge(gwas_ld_data,index_snps_simp[,1:2],by.x="SNP_A",by.y="SNP_A")

snp_gr <- GRanges(
  seqnames = index_snps_simp$CHR_A,
  ranges = IRanges(start = index_snps_simp$minimum, end = index_snps_simp$maximum),
)

mcols(snp_gr)$PEAK <- index_snps_simp$PEAK

# Perform the overlap
overlaps <- findOverlaps(snp_gr, annotations, ignore.strand = TRUE)

# Create a mapping table from overlaps
overlap_mapping <- data.frame(
  peak_index = queryHits(overlaps),
  annotation_index = subjectHits(overlaps)
)

overlap_mapping$PEAK <- mcols(snp_gr)$PEAK[overlap_mapping$peak_index]

# Convert annotations to a dataframe
overlapping_annotations_df <- as.data.frame(annotations)

# Ensure there's a unique identifier for each annotation
overlapping_annotations_df$unique_id <- seq_len(nrow(overlapping_annotations_df))

# Filter overlap_mapping to include only unique annotation indices
overlap_mapping_unique <- overlap_mapping[!duplicated(overlap_mapping$annotation_index), ]

# Merge SNP_A into overlapping_annotations_df
overlapping_annotations_df <- merge(overlapping_annotations_df, overlap_mapping_unique, by.x = "unique_id", by.y = "annotation_index", all.x = TRUE)
overlapping_annotations_df<-subset(overlapping_annotations_df,!is.na(PEAK))

gr=getBioMartGenes("pkingsleyae_gene_ensembl")

anno_meta<-as.data.frame(gr)[6:9]

overlapping_annotations_df<-left_join(overlapping_annotations_df,anno_meta,by=c("gene_id"="ensembl_gene_id"))
```


```{r}
fish_data_file<-file.path(root,"input_data/01_Terra/data_model/fish_data_2024a.txt")

fish_data<-read.csv(fish_data_file,row.names=1,sep="\t")

pheno_discrete <- fish_data %>% 
  mutate(Phenotype = factor(Phenotype)) %>%
  dplyr::select(Phenotype)

phenos<-setNames(as.character(pheno_discrete$Phenotype), row.names(pheno_discrete))
phenos <- gsub("TYPE_II", "TP", phenos)
phenos <- gsub("TYPE_I", "TP", phenos)
phenos<-as.factor(phenos)
## set colors for plotting 
col.pheno<-setNames(c("blue","red","purple"),levels(phenos)) 

phenos_df <- data.frame(Sample = names(phenos), Phenotype = as.factor(phenos))
```

## Plotting

### Make Genotype Matrix Plots
```{r}

tree_path=file.path("~/Desktop/for_github/pkings_gwas_2024_master/","output_data","04_Phylogenomics","pkings_iqtree_genes_astral_gcf_scf_tree.cf.tree.ultra.time.nex")
tree<-read.tree(tree_path)
tree<-drop.tip(tree,c("BIK_6927","COB_4006","PHOP_5511"))



# Perform a fuzzy join to match chr and position ranges (start to end)
df.merged <- long_genotype %>%
  fuzzy_inner_join(
    df.bed,
    by = c(
      "chr" = "chr",          # Match chromosomes
      "pos" = "start",        # Ensure pos is >= start
      "pos" = "end"           # Ensure pos is <= end
    ),
    match_fun = list(`==`, `>=`, `<=`)  # Match chr exactly, and pos within start and end
  ) %>%
  dplyr::select(-chr.y,-start,-end,-lg) %>%  # Explicitly drop chr.y
  dplyr::rename(chr = chr.x)  # Rename chr.x to chr


# Add SNP order within individual, grouped by peak and name
df.genos.sub <- df.merged %>%
  arrange(peak, name, pos) %>%
  group_by(peak, name) %>%
  mutate(snp_order_within_individual = row_number()) %>%
  ungroup()

# Convert geno to a factor with desired levels
df.genos.sub$geno <- factor(ifelse(is.na(df.genos.sub$geno), "NA", as.character(df.genos.sub$geno)),
                            levels = c("NA", "0", "1", "2"))

# Merge phenotype information
df.genos.sub <- df.genos.sub %>%
  mutate(population = sub("_.*", "", name)) %>%  # Extract population from the name
  left_join(phenos_df, by = c("name" = "Sample"))  # Merge phenos_df to include Phenotype

# Sort by phenotype within each population and reverse name order for plotting
df.genos.sub <- df.genos.sub %>%
  group_by(peak, population) %>%
  arrange(Phenotype, name, .by_group = TRUE) %>%
  ungroup() %>%
  mutate(name = factor(name, levels = rev(unique(name))))  # Reverse levels for top-to-bottom plotting

# Identify the first individual (top-to-bottom) for each population within each peak
first_individuals <- df.genos.sub %>%
  group_by(peak, population) %>%
  slice_head(n = 1) %>%  # Select the first individual per population
  dplyr::select(peak, name, population)  # Keep only peak, name, and population

# Create a mapping for y-axis labels: Label only the first individual per population
label_mapping <- setNames(first_individuals$population, first_individuals$name)

# Convert population to factor
df.genos.sub$population <- as.factor(df.genos.sub$population)
df.genos.sub$peak <- as.factor(df.genos.sub$peak)
df.genos.sub$name <- as.character(df.genos.sub$name)

```

```{r,fig.width=12}
# The output data frame now includes the peak information for faceting

  # Step 4: Create the ggplot object
  geno_plot<-create_phylo_geno_plot_multi(tree,df.genos.sub)
  
  print(geno_plot)
```


### Plot Boxplots of Popgen Stats by Peak
```{r}
pi_res<-get_data4plot(df.bed,pi_plot,'avg_pi')

pi_stats<-pi_res %>%
  group_by(peak ) %>%
  t_test(avg_pi ~ population ) %>%
adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")

td_res<-get_data4plot(df.bed,tajd_plot,'avg_tajd')

td_stats<-td_res %>%
  group_by(peak ) %>%
  t_test(avg_tajd ~ population ) %>%
adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")

fst_res<-get_fst_data4plot(df.bed,fst_plot)

fst_stats<-fst_res %>%
  group_by(peak ) %>%
  t_test(avg_wc_fst ~ comparison ) %>%
adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj")
```

```{r, fig.height=8}
custom_palette <- c(
  "APA" = "#1f78b4", "APAR" = "#a6cee3", # Blue hues for APA
  "BAM" = "#33a02c", "BAMR" = "#b2df8a", # Green hues for BAM
  "BEN" = "#e31a1c", "BENR" = "#fb9a99"  # Red hues for BEN
)

# Create the base plot with custom colors
bxp <- ggplot(pi_res, aes(x = peak, y = avg_pi)) +
  geom_boxplot(aes(fill = population), outlier.shape = NA) +
  scale_fill_manual(values = custom_palette) + # Apply custom palette
  coord_cartesian(ylim = c(quantile(pi_res$avg_pi, 0.01), quantile(pi_res$avg_pi, 0.99)))

# Filter and adjust statistical test data for the plot
stat.test4plot <- pi_stats %>%
  filter(p.adj.signif != "ns") %>%
  add_xy_position(x = "peak", dodge = 0.8, step.increase = 0.1, fun = "median_iqr")

# Add statistical annotations
bxp<-bxp + 
  stat_pvalue_manual(
    stat.test4plot,
    label = "p.adj.signif",
    tip.length = 0.002
  )+
theme_minimal()

# Create the base plot with custom colors
bxt <- ggplot(td_res, aes(x = peak, y = avg_tajd)) +
  geom_boxplot(aes(fill = population), outlier.shape = NA) +
  scale_fill_manual(values = custom_palette) #+ # Apply custom palette
  #coord_cartesian(ylim = c(quantile(td_res$avg_tajd, 0.01), quantile(td_res$avg_tajd, 0.99)))

# Filter and adjust statistical test data for the plot
stat.test4plot <- td_stats %>%
  filter(p.adj.signif != "ns") %>%
  add_xy_position(x = "peak", dodge = 0.8, step.increase = 0.1, fun = "median_iqr")

# Add statistical annotations
bxt<-bxt + 
  stat_pvalue_manual(
    stat.test4plot,
    label = "p.adj.signif",
    tip.length = 0.002
  )+
theme_minimal()

bxf <- ggplot(fst_res, aes(x = peak, y = avg_wc_fst)) +
  geom_boxplot(aes(fill = comparison), outlier.shape = NA)

stat.test4plot <- fst_stats %>%
  filter(p.adj.signif != "ns") %>%
  add_xy_position(x = "peak", dodge = 0.8, step.increase = 0.1, fun = "median_iqr")

bxf<-bxf + 
  stat_pvalue_manual(
    stat.test4plot,
    label = "p.adj.signif",
    tip.length = 0.002
  )+
theme_minimal()

combined_peak_plot <- bxp / bxt /bxf  +
    plot_annotation( tag_levels = 'A') &
    theme(plot.tag = element_text(size = 24, face = "bold"))
  
  # Adjust relative heights of the plots
  combined_peak_stats <- combined_peak_plot + plot_layout(heights = c(2, 2,2))

print(combined_peak_stats)

```

```{r}
window_size<-100000

peak.pk.lmm <- pk.lmm %>%
  inner_join(df.bed, by = c("numeric_chr" = "chr")) %>%
  filter(ps >= start-window_size & ps <= end+window_size)



colnames(gwas_ld_data)[colnames(gwas_ld_data) == "PEAK"] <- "peak"

peak.gwas_ld_data<-gwas_ld_data %>%
  inner_join(df.bed, by = c("peak" = "peak")) %>%
  filter(ps >= start-window_size & ps <= end+window_size)

colnames(overlapping_annotations_df)[colnames(overlapping_annotations_df) == "PEAK"] <- "peak"


snp_plot<-ggplot() +
  geom_point(data = peak.pk.lmm, mapping = aes(x = ps, y = -log10(p_lrt)), color = "grey")+
  geom_point(data = peak.gwas_ld_data, mapping = aes(x = BP_B, y = -log10(p_lrt), color = R2)) +
  scale_color_gradientn(colours = c("blue", "green", "yellow", "red")) +
  facet_wrap(~ peak,scales="free_x")+
  geom_hline(yintercept = significant, linetype = "dashed", color = "red", size = 0.5) +
  geom_hline(yintercept = suggestive, linetype = "dashed", color = "blue", size = 0.5) +
  geom_gene_arrow(data = overlapping_annotations_df, mapping = aes(xmin = start, xmax = end, y = -1, fill = external_gene_name), show.legend = FALSE) +
  geom_gene_label(data = overlapping_annotations_df, mapping = aes(xmin = start, xmax = end, y = -1, label = external_gene_name), align = "left", show.legend = FALSE)+
  theme_minimal() +
  theme(
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_text(size = rel(1.2), family = "Arial", angle = 25, hjust = 1),
    axis.text.y = element_text(size = rel(1.2), family = "Arial", angle = 0, hjust = 1),
    axis.title.x = element_text(size = rel(1.2), family = "Arial", margin = margin(t = 20)),
    axis.title.y = element_text(size = rel(1.2), family = "Arial", margin = margin(r = 20)),
    legend.position = "right"
  ) +
  xlab("Chromosome Position") +
  ylab("-log(P)")

snp_plot
```

## Final Plot
```{r,fig.width=12}

thebigplot <- wrap_plots(
  (wrap_plots(snp_plot, geno_plot, ncol = 1, heights = c(0.5, 2))), # Left stack
  (wrap_plots(bxp, bxt, bxf, ncol = 1)),                         # Right stack
  widths = c(2, 1)                                               # Left vs. right proportions
) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 12, face = "bold"))

print(thebigplot)
```


