---
title: "Peak Report"
output: html_notebook
params:
  data_path: PKINGS_ALL_WOB_EXCLUDED
  output_file: NULL
  popgen_prefix: MIC4273_HAP2_NEWREF_MSU618_all_fish_all_sites
  MAF_thresh: 0.15
  pops: ["APA","BAM","BEN"]
  phenos: ["TP","BP","TP"]
---
#data_path: MOV_BIR_DOG_DOV_BAVA_BIK_APA_BEN_TP1_BP2_WOB9
#pops: ["MOV","BIR","DOG","BAVA","BIK","APA","BEN"]
#phenos: ["TP","BP","TP","BP","BP","TP","TP"]
#MAF_thresh: 0.15

#data_path: APA_BEN_BAM_TP1_BP2_WOB9
#pops: ["APA","BAM","BEN"]
#phenos: ["TP","BP","TP"]
#MAF_thresh: 0.05

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(tidyverse)
library(patchwork)
library(data.table)
library(reshape2)
library(stringr)
library(scales)      # For percent_format in scale_y_continuous
require(GenomicRanges)
require(rtracklayer)
require(gggenes)
require(rentrez)
require(rlang)
require(biomaRt)
library(BioMartGOGeneSets)
library(ggtree)
library(ape)
library(ggsignif)
library(fuzzyjoin)
library(rstatix)
library(ggpubr)
library(RColorBrewer)
library(ggrepel)
library(flextable)
library(officer)
library(lme4)
library(lmerTest)
library(purrr)
library(broom.mixed)

source(file.path(root,"code","06_Association","gemma_gwas_functions.R"))
source(file.path(root,"code","08_Peak_Analysis","peak_analysis_functions.R"))
source(file.path(root,"code","08_Peak_Analysis","phylogenoplot_multi.R"))

rename.tips <- function(phy, old_names, new_names) {
   mpos <- match(old_names,phy$tip.labels)
   phy$tip.labels[mpos] <- new_names
   return(phy)
}

```

## Data Loading

### Load the Genotypic Data
```{r}
geno_path<-file.path(root,"output_data","08_Peak_Analysis",params$data_path,paste0(params$data_path,".bin.pos.genos"))
peaks_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_PEAKS.bed"))
logp_path<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_SNPS_IN_PEAKS_TOP250.txt"))
dxy_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_dxy.txt"))
fst_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_fst.txt"))
pi_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_pi.txt"))
tajd_file=file.path(root,"output_data/05_PopGen/",paste0(params$popgen_prefix,"_tajd.txt"))


df.genos<-read.table(geno_path,header=T)

long_genotype <- df.genos %>%
  pivot_longer(
    cols = -c(CHROM, POS, REF, ALT), # Exclude the first four columns
    names_to = "name",          # Column for the new 'name' variable
    values_to = "geno"          # Column for the new 'geno' variable
  ) %>%
  dplyr::select(name, chr = CHROM, pos = POS, geno) # Rename and select columns

long_genotype<-as.data.frame(long_genotype)

long_genotype<-long_genotype %>% mutate(chr=str_replace(chr,"chr",""))

df.bed <- read.table(peaks_path)
names(df.bed) <- c("chr", "start","end", "peak","maxp","maxp_pos")

#remove positions if there is only one SNP
df.bed <- df.bed %>% 
  mutate(lg = end- start) %>% 
  filter(lg > 0)

#load the top p values (filtered > 0.1) to plot the signal with the other plots I generate below
df.logp <- read.table(logp_path)

```

### Load the Population Genetics Statistics
```{r}

pixy_df <- list()

tajd_df<-fread(tajd_file)

tajd_df<-tajd_df %>% arrange(pop1,CHROM,BIN_START)

tajd_df <- tajd_df %>%
    mutate(window_pos_1=BIN_START+1) %>%
    mutate(window_pos_2=BIN_START+5000) %>%
    mutate(chromosome=CHROM) %>%
    mutate(statistic="avg_tajd") %>%
    mutate(value=TajimaD) %>%
    mutate(pop2=NA) %>%
    dplyr::select(pop1,  pop2, chromosome, window_pos_1, window_pos_2, statistic, value, ) %>%
    arrange(pop1,  pop2, chromosome, window_pos_1)

df_pi <- fread(pi_file)
      
df_pi <- df_pi %>%
  gather(-pop, -window_pos_1, -window_pos_2, -chromosome,
  key = "statistic", value = "value") %>%
  mutate(pop1= pop) %>%
  mutate(pop2 = NA) %>%
  dplyr::select(pop1, chromosome, window_pos_1, window_pos_2, statistic, value, pop2 )

fst_df <- fread(fst_file)

fst_df <- fst_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
dxy_df <- fread(dxy_file)

dxy_df <- dxy_df %>%
        gather(-pop1, -pop2, -window_pos_1, -window_pos_2, -chromosome,
               key = "statistic", value = "value")
      
      
pixy_df[[1]] <- fst_df
pixy_df[[2]] <- dxy_df
pixy_df[[3]] <- df_pi
pixy_df[[4]] <- tajd_df

pixy<-bind_rows(pixy_df) %>%
    arrange(pop1, pop2, chromosome, window_pos_1, statistic)


pixy<-pixy %>% pivot_wider(names_from=statistic,values_from=value) %>% mutate(comparison=paste0(pop1,"_",pop2))

pixy <- pixy %>% 
  pivot_longer(cols = -c(pop1, pop2, chromosome, window_pos_1, window_pos_2,comparison),
               names_to = "statistic", values_to = "value")

rm(df_pi,dxy_df,fst_df,pixy_df,tajd_df)
gc()

```

### Filter Population Genetics Statistics
```{r}
phenotypes<-data.frame(POP=params$pops,population_phenotype=params$phenos)

# Dynamically create the filter condition
filter_expr <- paste0("pop1 == '", params$pops, "'", collapse = " | ")

# Generate all pairwise combinations
pairs <- combn(params$pops, 2, simplify = FALSE)

# Create forward comparisons
forward <- sapply(pairs, function(x) paste(x, collapse = "_"))

# Create backward comparisons
backward <- sapply(pairs, function(x) paste(rev(x), collapse = "_"))

# Combine forward and backward comparisons
comparison_list <- c(forward, backward)

comparision_expr <- paste0("comparison == '", comparison_list, "'", collapse = " | ")

pi_plot <- pixy %>%
  filter(statistic == "avg_pi", !is.na(value)) %>%
  filter(!!rlang::parse_expr(filter_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(phenotype_pop1 = population_phenotype)

tajd_plot<-pixy %>%
    filter(statistic == "avg_tajd", !is.na(value)) %>%
    filter(!!rlang::parse_expr(filter_expr)) %>%
    mutate(value = as.numeric(value) ) %>%  
    left_join(phenotypes, by = c("pop1" = "POP")) %>%
    dplyr::rename(phenotype_pop1 = population_phenotype)

fst_plot <- pixy %>%
  filter(statistic == "avg_wc_fst", !is.na(value)) %>%
  filter(!!rlang::parse_expr(comparision_expr)) %>%
  mutate(value = as.numeric(value)) %>%
  left_join(phenotypes, by = c("pop1" = "POP")) %>%
  dplyr::rename(pop1_phenotype = population_phenotype) %>%
  left_join(phenotypes, by = c("pop2" = "POP")) %>%
  dplyr::rename(pop2_phenotype = population_phenotype) %>%
  mutate(
    comparison_classification = case_when(
      pop1_phenotype == pop2_phenotype ~ "Same phenotype",
      pop1_phenotype != pop2_phenotype ~ "Different phenotypes",
      TRUE ~ "Unknown"
    )) %>%
  mutate(
    compare_color = case_when(
      comparison_classification == "Same phenotype" & pop1_phenotype=="BP" ~ "BP-BP",
      comparison_classification == "Same phenotype" & pop1_phenotype=="TP" ~ "TP-TP",
      comparison_classification == "Different phenotypes" ~ "BP-TP"
    )
  )
  

```

### Load SNP Data
```{r}
anno_file<-file.path(root,"input_data","00_Reference_Genome","jordan-mic4273-mb-hirise-20xx3__06-13-2024__final_assembly.ensembl.liftoff.renamed.genes_only.gff")
gemma_file<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,".lmm2.assoc.txt"))
minpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_permution"),"min.txt")
ldpath<-file.path(root,"output_data","06_Association",params$data_path,paste0(params$data_path,"_index_snps.ld.ld"))
peak_data_path<-file.path(root,"output_data","06_Association",params$data_path)


#start with loading the association data
pk.lmm<- gemma.order(gemma_file, "p_lrt")
min_data<-read.csv(minpath,header=F,sep="\t")
pvals<-sort(min_data$V10, decreasing = F)
significant = -log10(pvals[length(pvals) - floor(length(pvals)*0.95)])
suggestive = -log10(pvals[length(pvals) - floor(length(pvals)*0.67)])

#remove low log p to reduce memory footprint
pk.lmm.filt <- pk.lmm %>% filter(log_p > 1) %>% filter(af > params$MAF_thresh) %>% filter((!grepl("ups", chr)) & !is.na(row) & !is.na(log_p)) %>% arrange(numeric_chr)
pk.lmm.filt<- pk.lmm.filt %>% mutate(new_rs = str_split(rs, ":", simplify = TRUE)[, 1])

# now read the ld data
ld_data<-fread(ldpath)

#Merge the LD Data and the Association Data
gwas_ld_data<-left_join(pk.lmm.filt,ld_data,by=c("new_rs"="SNP_B"))

# Get the Data inside the Peaks
pk.lmm.peaks.out<-read.table(file.path(peak_data_path,paste0(params$data_path,"_SNPS_IN_PEAKS.txt")),header=T)
pk.lmm.peaks.out$rs<-str_split(pk.lmm.peaks.out$rs,":",simplify=TRUE)[,1]


## Identify the "Index" SNPS (SNP with Highest P-value with associated R^2 Value)
# Function to get index SNPs with >1 R² value
get_index_snps <- function(peaks_data, ld_data) {
  peaks_data %>%
    group_by(peak) %>%
    arrange(desc(log_p)) %>%  # Order by descending log_p within each peak
    mutate(has_ld = map_lgl(rs, ~ {
      # Check if the SNP has >1 associated SNP in ld_data with R^2 values
      snp_data <- ld_data %>% filter(SNP_B == .x)
      nrow(snp_data) > 1
    })) %>%
    filter(has_ld) %>%        # Keep only rows where SNP has associated R^2 values
    slice_head(n = 1) %>%     # Take the first valid SNP per peak
    ungroup()
}

# Call the function to identify valid index SNPs
index_snps <- get_index_snps(pk.lmm.peaks.out, ld_data)

# Simplify the output
index_snps_simp <- index_snps[, c(16, 2, 1, 3, 10)]
colnames(index_snps_simp) <- c("PEAK", "SNP_A", "CHR_A", "BP_A", "p_lrt")

# Define Windows for Display / Analysis
index_snps_simp$maximum <- index_snps_simp$BP_A + 50000
index_snps_simp$minimum <- index_snps_simp$BP_A - 50000

# One of the SNPs is near an assembly gap, so let's manually adjust that window to see it.
index_snps_simp <- index_snps_simp %>%
  mutate(
    maximum = ifelse(SNP_A == "chr8_15781018_T_A", 16.1e6, maximum),
    minimum = ifelse(SNP_A == "chr8_15781018_T_A", 15.65e6, minimum)
  )

index_snps_simp <- index_snps_simp %>%
  mutate(
    minimum = ifelse(SNP_A == "chr16_10915940_C_A", 10.84e6, minimum),
    maximum = ifelse(SNP_A == "chr16_10915940_C_A", 11.04e6, maximum)
  )

index_snps_simp <- index_snps_simp %>%
  mutate(
    minimum = ifelse(SNP_A == "chr16_10916883_A_G", 10.84e6, minimum),
    maximum = ifelse(SNP_A == "chr16_10916883_A_G", 11.04e6, maximum)
  )


#Keep LD Values only associated with the index SNP
gwas_and_ld_data_only<-left_join(index_snps_simp[,1:2],gwas_ld_data, by=("SNP_A"="SNP_A"))

gwas_and_ld_data_only_new <- map(1:nrow(index_snps_simp), function(i) {
  # Identify the boundaries
  left_bound  <- index_snps_simp$minimum[i]
  right_bound <- index_snps_simp$maximum[i]
  
  # Filter pk.lmm by chromosome and position within the window
  gwas_and_ld_data_only %>%
    filter(
      chr == index_snps_simp$CHR_A[i],
      ps >= left_bound,
      ps <= right_bound
    ) 
})

pk.lmm.filt_new<-map(1:nrow(index_snps_simp), function(i) {
  # Identify the boundaries
  left_bound  <- index_snps_simp$minimum[i]
  right_bound <- index_snps_simp$maximum[i]
  
  # Filter pk.lmm by chromosome and position within the window
  pk.lmm.filt %>%
    filter(
      chr == index_snps_simp$CHR_A[i],
      ps >= left_bound,
      ps <= right_bound
    ) %>% mutate(
      PEAK=index_snps_simp$PEAK[i]
    )
})

gwas_and_ld_data_only_new <- bind_rows(gwas_and_ld_data_only_new)

pk.lmm.filt_new <-bind_rows(pk.lmm.filt_new)

annotations<-import(anno_file)

#Define a GRANGES to intersect with the annotations data
snp_gr <- GRanges(
  seqnames = index_snps_simp$CHR_A,
  ranges = IRanges(start = index_snps_simp$minimum, end = index_snps_simp$maximum),
)

mcols(snp_gr)$PEAK <- index_snps_simp$PEAK

# Perform the overlap
overlaps <- findOverlaps(snp_gr, annotations, ignore.strand = TRUE)

# Create a mapping table from overlaps
overlap_mapping <- data.frame(
  peak_index = queryHits(overlaps),
  annotation_index = subjectHits(overlaps)
)

overlap_mapping$PEAK <- mcols(snp_gr)$PEAK[overlap_mapping$peak_index]

# Convert annotations to a dataframe
overlapping_annotations_df <- as.data.frame(annotations)

# Ensure there's a unique identifier for each annotation
overlapping_annotations_df$unique_id <- seq_len(nrow(overlapping_annotations_df))

# Filter overlap_mapping to include only unique annotation indices
overlap_mapping_unique <- overlap_mapping[!duplicated(overlap_mapping$annotation_index), ]

# Merge SNP_A into overlapping_annotations_df
overlapping_annotations_df <- merge(overlapping_annotations_df, overlap_mapping_unique, by.x = "unique_id", by.y = "annotation_index", all.x = TRUE)
overlapping_annotations_df<-subset(overlapping_annotations_df,!is.na(PEAK))

gr=getBioMartGenes("pkingsleyae_gene_ensembl")

anno_meta<-as.data.frame(gr)[6:9]

overlapping_annotations_df<-left_join(overlapping_annotations_df,anno_meta,by=c("gene_id"="ensembl_gene_id"))

overlapping_annotations_for_output<-overlapping_annotations_df[c(2,3,4,11,13,27)]
write.table(overlapping_annotations_for_output,file.path(peak_data_path,paste0(params$data_path,"_GENES_IN_PEAKS.txt")), sep = "\t", quote = F, row.names = F, col.names = T)
```


```{r}
fish_data_file<-file.path(root,"input_data/01_Terra/data_model/fish_data_2024a.txt")

fish_data<-read.csv(fish_data_file,row.names=1,sep="\t")

pheno_discrete <- fish_data %>% 
  mutate(Phenotype = factor(Phenotype)) %>%
  dplyr::select(Phenotype)

phenos<-setNames(as.character(pheno_discrete$Phenotype), row.names(pheno_discrete))
phenos <- gsub("TYPE_II", "TP", phenos)
phenos <- gsub("TYPE_I", "TP", phenos)
phenos<-as.factor(phenos)
## set colors for plotting 
col.pheno<-setNames(c("blue","red","purple"),levels(phenos)) 

phenos_df <- data.frame(Sample = names(phenos), Phenotype = as.factor(phenos))
```

## Plotting

### Make Genotype Matrix Plots
```{r}

tree_path=file.path("~/Desktop/for_github/pkings_gwas_2024_master/","output_data","04_Phylogenomics","pkings_iqtree_genes_astral_gcf_scf_tree.cf.tree.ultra.time.nex")
tree<-read.tree(tree_path)
tree<-drop.tip(tree,c("BIK_6927","COB_4006","PHOP_5511"))



# Perform a fuzzy join to match chr and position ranges (start to end)
df.merged <- long_genotype %>%
  fuzzy_inner_join(
    df.bed,
    by = c(
      "chr" = "chr",          # Match chromosomes
      "pos" = "start",        # Ensure pos is >= start
      "pos" = "end"           # Ensure pos is <= end
    ),
    match_fun = list(`==`, `>=`, `<=`)  # Match chr exactly, and pos within start and end
  ) %>%
  dplyr::select(-chr.y,-start,-end,-lg) %>%  # Explicitly drop chr.y
  dplyr::rename(chr = chr.x)  # Rename chr.x to chr


# Add SNP order within individual, grouped by peak and name
df.genos.sub <- df.merged %>%
  arrange(peak, name, pos) %>%
  group_by(peak, name) %>%
  mutate(snp_order_within_individual = row_number()) %>%
  ungroup()

# Convert geno to a factor with desired levels
df.genos.sub$geno <- factor(ifelse(is.na(df.genos.sub$geno), "NA", as.character(df.genos.sub$geno)),
                            levels = c("NA", "0", "1", "2"))

# Merge phenotype information
df.genos.sub <- df.genos.sub %>%
  mutate(population = sub("_.*", "", name)) %>%  # Extract population from the name
  left_join(phenos_df, by = c("name" = "Sample"))  # Merge phenos_df to include Phenotype

# Sort by phenotype within each population and reverse name order for plotting
df.genos.sub <- df.genos.sub %>%
  group_by(peak, population) %>%
  arrange(Phenotype, name, .by_group = TRUE) %>%
  ungroup() %>%
  mutate(name = factor(name, levels = rev(unique(name))))  # Reverse levels for top-to-bottom plotting

# Identify the first individual (top-to-bottom) for each population within each peak
first_individuals <- df.genos.sub %>%
  group_by(peak, population) %>%
  slice_head(n = 1) %>%  # Select the first individual per population
  dplyr::select(peak, name, population)  # Keep only peak, name, and population

# Create a mapping for y-axis labels: Label only the first individual per population
label_mapping <- setNames(first_individuals$population, first_individuals$name)

# Convert population to factor
df.genos.sub$population <- as.factor(df.genos.sub$population)
df.genos.sub$peak <- as.factor(df.genos.sub$peak)
df.genos.sub$name <- as.character(df.genos.sub$name)

```

```{r,fig.width=12}
# The output data frame now includes the peak information for faceting

  # Step 4: Create the ggplot object
  geno_plot<-create_phylo_geno_plot_multi(tree,df.genos.sub,peaks_per_plot = 4)
  
  print(geno_plot)
```


### Plot Boxplots of Popgen Stats by Peak
```{r}
df.bed <- df.bed[order(-df.bed$maxp), ]
pi_res<-get_data4plot(df.bed,pi_plot,'avg_pi')
td_res<-get_data4plot(df.bed,tajd_plot,'avg_tajd')
fst_res<-get_fst_data4plot(df.bed,fst_plot)
```

```{r, fig.height=12}

phenotype_palettes <- list(
  "BP" = "Blues",    # Blue shades for BP
  "TP" = "Reds",     # Red shades for TP
  "Wobble" = "Purples" # Purple shades for Wobble
)

custom_palette <- generate_palette_alpha(params$pops, params$phenos, phenotype_palettes)

pw_fst_comparisions<-fst_plot %>% dplyr::select(comparison,compare_color) %>% unique()

# Define compare_colors mapping
compare_colors <- list(
  "BP-BP" = "Greys",
  "TP-TP" = "Greys",
  "BP-TP" = "Reds"
)

fst_palette <- generate_fst_palette_alpha(pw_fst_comparisions, compare_colors)

peaks_per_plot <- 4

# Extract unique peaks and split into groups

  ordered_levels <- df.bed$peak[order(-df.bed$maxp)]
  df.bed$peak <- factor(df.bed$peak, levels = unique(ordered_levels))
  peak_levels <- levels(df.bed$peak)
  peak_levels <- paste0("peak",peak_levels)
  peak_groups <- split(peak_levels, ceiling(seq_along(peak_levels) / peaks_per_plot))



bxp_plot_list <- list()  # Initialize an empty list to store plots


for (group in peak_groups) {
  # Filter data for the current group of peaks
  pi_res_group <- pi_res %>% filter(paste0("peak",peak) %in% group)
  td_res_group <- td_res %>% filter(paste0("peak",peak) %in% group)
  fst_res_group <- fst_res %>% filter(paste0("peak",peak) %in% group)

  pi_res_group$peak<-paste0("peak",pi_res_group$peak)
  td_res_group$peak<-paste0("peak",td_res_group$peak)
  fst_res_group$peak<-paste0("peak",fst_res_group$peak)

  # Convert `peak` to a factor with levels restricted to the current group
  pi_res_group <- pi_res_group %>% mutate(peak = factor(peak, levels = group))
  td_res_group <- td_res_group %>% mutate(peak = factor(peak, levels = group))
  fst_res_group <- fst_res_group %>% mutate(peak = factor(peak, levels = group))
  
  # plot pi
  pi_res_group <- pi_res_group %>%
    mutate(category = if_else(population %in% params$pops, "peak", "random"))
  
  pi_res_group <- pi_res_group %>%
  mutate(population = if_else(
    category == "random",
    str_replace(population, "R$", ""),  # Remove terminal 'R' if it exists
    population
  ))
  
  pheno_lut<-pi_plot %>% dplyr::select(pop1,phenotype_pop1) %>% unique()
  
  pi_res_group<-pi_res_group %>% left_join(pheno_lut,by=join_by(population==pop1))
  
  # Create the base plot
  bxp <- ggplot(pi_res_group, aes(x = population, y = avg_pi)) +
  geom_boxplot(aes(fill = phenotype_pop1, alpha=category), outlier.shape = NA) +
  facet_wrap(~peak, ncol = 4) +
  scale_fill_manual(values = c("BP" = "blue", "TP" = "red")) +  # Custom fill colors
    scale_alpha_manual(
    values = c("random" = 0.2, "peak" = 1.0)  # Set alpha based on category
  ) +
  labs(y = expression("π")) +
  theme_minimal() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.grid.minor.y = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text = element_text(size = rel(1.2), family = "Arial", hjust = 1),
    axis.title.x = element_text(size = rel(1.2), family = "Arial", margin = margin(t = 20)),
    axis.title.y = element_text(size = rel(1.2), family = "Arial", margin = margin(r = 20))
  ) +
  coord_cartesian(ylim = c(quantile(pi_res_group$avg_pi, 0.01), quantile(pi_res_group$avg_pi, 0.98)))
  
  pi_res_group$phenotype_pop1 <- factor(pi_res_group$phenotype_pop1, levels = c("TP", "BP"))
  pi_res_group$category<-factor(pi_res_group$category,levels=c("random","peak"))

  pi_model <- lmer(avg_pi ~ category * phenotype_pop1 + (1| population) + (1 | peak) + (1 | interval), data=pi_res_group)
  model_summary <- capture.output(summary(pi_model))
  
  #print(model_summary)
  #writeLines(model_summary, con = file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,"_pi_model",paste(group,collapse="_"),".txt")))

  #plot tajim'as d
  td_res_group <- td_res_group %>%
    mutate(category = if_else(population %in% params$pops, "peak", "random"))
  
  td_res_group <- td_res_group %>%
  mutate(population = if_else(
    category == "random",
    str_replace(population, "R$", ""),  # Remove terminal 'R' if it exists
    population
  ))
  
  td_res_group<-td_res_group %>% left_join(pheno_lut,by=join_by(population==pop1))
  
  bxt <- ggplot(td_res_group, aes(x = population, y= avg_tajd)) +
  geom_boxplot(aes(fill = phenotype_pop1, alpha=category), outlier.shape = NA) +
  scale_fill_manual(values = c("BP" = "blue", "TP" = "red")) +  # Custom fill colors
    scale_alpha_manual(
    values = c("random" = 0.2, "peak" = 1.0)  # Set alpha based on category
  ) +
  labs(y=expression("Tajima's D"))+
    facet_wrap(~peak, ncol = 4) +
    
  #coord_cartesian(ylim = c(quantile(td_res$avg_tajd, 0.01), quantile(td_res$avg_tajd, 0.99)))
  theme_minimal()+
  theme(
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      axis.text.x = element_text(angle = 45, hjust = 1),
      axis.text = element_text(size = rel(1.2), family = "Arial",  hjust = 1),
      axis.title.x = element_text(size = rel(1.2), family = "Arial", margin = margin(t = 20)),
      axis.title.y = element_text(size = rel(1.2), family = "Arial", margin = margin(r = 20))
    )
  
  td_res_group$phenotype_pop1 <- factor(td_res_group$phenotype_pop1, levels = c("TP", "BP"))
  td_res_group$category<-factor(td_res_group$category,levels=c("random","peak"))

  td_model <- lmer(avg_tajd ~ category * phenotype_pop1 + (1 | population) + (1 | peak) + (1 | interval), data=td_res_group)
  model_summary <- capture.output(summary(td_model))
  
  #print(model_summary)
  
  #writeLines(model_summary, con = file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,"_tajimad_model",paste(group,collapse="_"),".txt")))
  
  
# plot fst

    compare_lut<-fst_plot %>% dplyr::select(comparison,compare_color) %>% unique()
    
    fst_res_group <- fst_res_group %>%
      left_join(compare_lut, by = "comparison")
    
    compare_lut_2<-fst_res_group %>% dplyr::select(comparison,compare_color) %>% unique() %>% filter(!is.na(compare_color))
    
    fst_res_group <- fst_res_group %>%
      mutate(type = if_else(
        sapply(strsplit(comparison, "_"), function(x) all(x %in% params$pops)),
        "peak",
        "random"
      ))
    
    fst_res_group <- fst_res_group %>%
      mutate(comparison = if_else(
        type == "random",
        str_replace(comparison, "R$", ""),  # Remove terminal 'R' if it exists
        comparison
      ))
    
    fst_res_group<- fst_res_group %>% dplyr::select(-compare_color)
    fst_res_group<-fst_res_group %>% left_join(compare_lut_2, by="comparison")
    #fst_res_group<-fst_res_group %>% filter(compare_color=="BP-TP")
    
    bxf<-ggplot(subset(fst_res_group,compare_color=="BP-TP"), aes(x = comparison, y = avg_wc_fst, fill = compare_color,alpha=type)) +
      geom_boxplot(outlier.shape = NA) +  # Narrower box plots
      facet_wrap(~ peak, scales = "free_x",ncol=4) +
      scale_fill_manual(values = c("BP-BP" = "blue", "BP-TP" = "purple", "TP-TP" = "red")) +  # Custom fill colors
      scale_alpha_manual(
        values = c("random" = 0.2, "peak" = 1.0)  # Set alpha based on category
      ) +
      coord_cartesian(ylim = c(0, 1)) +  # Fixed y-axis limits between 0 and 1
      theme_minimal() +
      labs(fill = "Type",
           color = "Type",
           shape = "Type",
           x = "Comparision",
           y = expression(F[st]))+
      theme(
      panel.grid.major.x = element_blank(),
          panel.grid.minor.x = element_blank(),
          axis.text.x = element_text(angle = 45, hjust = 1),
          axis.text = element_text(size = rel(1.2), family = "Arial",  hjust = 1),
          axis.title.x = element_text(size = rel(1.2), family = "Arial", margin = margin(t = 20)),
          axis.title.y = element_text(size = rel(1.2), family = "Arial", margin = margin(r = 20))
    )
    
    fst_res_group$compare_color <- factor(fst_res_group$compare_color, levels = c("TP-TP","BP-BP","BP-TP"))
    fst_res_group$type<-factor(fst_res_group$type,levels=c("random","peak"))
    fst_model <- lmer(avg_wc_fst ~ type * compare_color + (1 | comparison) + (1 | peak) + (1 | interval), data=fst_res_group)
    model_summary <- capture.output(summary(fst_model))
  
  #print(model_summary)
  
  #writeLines(model_summary, con = file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,"_fst_model",paste(group,collapse="_"),".txt")))

    combined_peak_plot <- bxp / bxt / bxf  +
    plot_annotation( tag_levels = 'A') &
    theme(plot.tag = element_text(size = 24, face = "bold"))
  # Add the plot to the list
  bxp_plot_list <- c(bxp_plot_list, list(combined_peak_plot))
  
  model_list <- list(
  "avg_pi" = pi_model, 
  "avg_tajd" = td_model, 
  "avg_wc_fst" = fst_model
)
  model_output <- map_df(model_list, 
                        ~ tidy(.x, effects = c("fixed","ran_pars")), 
                        .id = "model")

print(model_output)
  write.csv(model_output, file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,"_model_output_",paste(group,collapse="_"),".txt")),quote = F,row.names = F)

}



# Print each plot in the list
for (plot in bxp_plot_list) {
  print(plot)
}

```


```{r, fig.width=12}

colnames(overlapping_annotations_df)[colnames(overlapping_annotations_df) == "peak"] <- "PEAK"


maxp<-max(-log10(pk.lmm.filt_new$p_lrt))

# Replace NA values in `external_gene_name` with `gene_id`
overlapping_annotations_df <- overlapping_annotations_df %>%
  mutate(label_name = ifelse(is.na(external_gene_name), gene_id, external_gene_name)) %>%
  mutate(label_direction = ifelse(row_number() %% 2 == 0, "up", "down"))  # Alternate "up" and "down"

peaks_per_plot <- 4

# Extract unique peaks and split into groups
  ordered_levels <- index_snps_simp$PEAK[order(index_snps_simp$p_lrt)]
  index_snps_simp$PEAK <- factor(index_snps_simp$PEAK, levels = unique(ordered_levels))
  peak_levels <- levels(index_snps_simp$PEAK)
  peak_levels <- paste0("peak",peak_levels)
  peak_groups <- split(peak_levels, ceiling(seq_along(peak_levels) / peaks_per_plot))
  
snp_plot_list <- list()  # Initialize an empty list to store plots

for (group in peak_groups) {
  
  peak.pk.lmm_group <- pk.lmm.filt_new %>% filter(paste0("peak",PEAK) %in% group)
  peak.gwas_ld_data_group <- gwas_and_ld_data_only_new %>% filter(paste0("peak",PEAK) %in% group)
  overlapping_annotations_df_group <- overlapping_annotations_df %>% filter(paste0("peak",PEAK) %in% group)
  
  overlapping_annotations_df_group <- overlapping_annotations_df_group %>%
  mutate(
    forward = case_when(
      strand == "+" ~ TRUE,
      strand == "-" ~ FALSE,
      TRUE          ~ NA
    )
  ) %>%
  mutate(
    ypos = case_when(
      strand == "+" ~ (maxp + 3),  # shift slightly upward
      strand == "-" ~ (maxp + 1),  # shift slightly downward
      TRUE          ~ (maxp + 3)         # in case of '*'
    )
  )
  
  snp_plot <- ggplot() +
    geom_point(data = peak.pk.lmm_group, mapping = aes(x = ps / 1e6, y = -log10(p_lrt)), color = "grey") +
    geom_point(data = peak.gwas_ld_data_group, mapping = aes(x = BP_B / 1e6, y = -log10(p_lrt), color = R2)) +
  scale_color_gradientn(
      colours = rev(brewer.pal(9, "RdYlBu")),  # Reverse the palette for intuitive hot-to-cold mapping
      name = expression(R^2)                  # Label the legend for R²
    ) +
    scale_x_continuous(expand = c(0.01, 0), name = "Chromosome Position (Mb)",  breaks = scales::breaks_extended(n = 4)) +
    facet_wrap(~ PEAK, scales = "free_x",ncol=4) +
    geom_hline(yintercept = significant, linetype = "dashed", color = "red", size = 0.5) +
    geom_hline(yintercept = suggestive, linetype = "dashed", color = "blue", size = 0.5) +
    geom_gene_arrow(data = overlapping_annotations_df_group, 
                    mapping = aes(xmin = start / 1e6, xmax = end / 1e6, y = ypos,forward = forward),
                    fill = "darkgreen",
                    alpha=0.4,
                    show.legend = FALSE) +
    # Upward-moving labels
  geom_text_repel(
    data = overlapping_annotations_df_group %>% filter(label_direction == "up"),
    aes(x = (start / 1e6 + end / 1e6) / 2, y = ypos, label = label_name),
    nudge_y = 1,  # Move upward
    inherit.aes = FALSE,
    arrow = arrow(length = unit(0.1, "cm")),
    segment.color = "black",
    size = 5,
    force = 2,
    max.overlaps = Inf,
    box.padding = 0.6,
    point.padding = 0.3
  ) +
  # Downward-moving labels
  geom_text_repel(
    data = overlapping_annotations_df_group %>% filter(label_direction == "down"),
    aes(x = (start / 1e6 + end / 1e6) / 2, y = ypos-1, label = label_name),
    nudge_y = -1,  # Move downward
    inherit.aes = FALSE,
    arrow = arrow(length = unit(0.1, "cm")),
    segment.color = "black",
    size = 5,
    force = 2,
    max.overlaps = Inf,
    box.padding = 0.5,
    point.padding = 0.3
  )+
    theme_minimal() +
    theme(
      strip.background = element_blank(), 
            #strip.text = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank(),
      panel.grid.minor.y = element_blank(),
      axis.line.x = element_line(color = "black", size = 0.5),
      axis.ticks.x = element_line(color = "black", size = 0.5),
      axis.text.x = element_text(size = rel(2), family = "Arial", angle = 90, hjust = 1),
      axis.text.y = element_text(size = rel(2), family = "Arial", angle = 0, hjust = 1),
      axis.title.x = element_text(size = rel(2), family = "Arial", margin = margin(t = 20)),
      axis.title.y = element_text(size = rel(2), family = "Arial", margin = margin(r = 20)),
      legend.position = "right",
      legend.text = element_text(size=rel(1.2), family = "Arial"),
      panel.spacing = unit(2, "lines")
    ) +
    labs(y = expression("-log"[10] * " p-value"))
  
  # Add the plot to the list
  snp_plot_list <- c(snp_plot_list, list(snp_plot))
}

print(snp_plot_list)
```
```{r, fig.width=24}
# Assuming snp_plot, geno_plot, and combined_peak_stats are lists of equal length
thebigplot_list <- list()

for (i in seq_along(snp_plot_list)) {
  thebigplot_list[[i]] <- wrap_plots(
    (wrap_plots(snp_plot_list[[i]], geno_plot[[i]], ncol = 1, heights = c(1, 2))), # Left stack
    widths = 1                                                            # Left vs. right proportions
  ) +
    plot_annotation(tag_levels = 'A') &
    theme(plot.tag = element_text(size = 24, face = "bold"))
}

for (i in seq_along(snp_plot_list)) {
  ggsave(file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,".peak_analysis_",i,".svg")),plot=thebigplot_list[[i]],limitsize = FALSE,width = 24, height=14)
}

# Save or print the plots
print(thebigplot_list)

```


```{r, fig.width=14}

# Assuming snp_plot, geno_plot, and combined_peak_stats are lists of equal length
thebigplot_list <- list()

for (i in seq_along(snp_plot_list)) {
  thebigplot_list[[i]] <- wrap_plots(
    bxp_plot_list[[i]],                                                   # Right stack
    widths = 1                                                            # Left vs. right proportions
  ) +
    plot_annotation(tag_levels = 'A') &
    theme(plot.tag = element_text(size = 24, face = "bold"))
}

for (i in seq_along(snp_plot_list)) {
  ggsave(file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,".popgen_",i,".svg")),plot=thebigplot_list[[i]],limitsize = FALSE,width = 8, height=14)
}

# Save or print the plots
print(thebigplot_list)
```



```{r, echo = FALSE, results='asis'}

ensembl_112 <- useEnsembl(biomart = "ensembl", version = 112,host="https://may2024.archive.ensembl.org")
ensembl_112 <-useDataset("pkingsleyae_gene_ensembl",mart=ensembl_112)

# Assuming the attribute 'description' exists, retrieve the gene descriptions
gene_descriptions <- getBM(values=overlapping_annotations_df$gene_id,
                                filters = "ensembl_gene_id",
  attributes = c("ensembl_gene_id", "description"),
  mart = ensembl_112
)


suggestive_snps<-NULL
snp_count_per_peak<-NULL
genes_per_peak<-NULL
final_report<-NULL
kable_table<-NULL

gwas_and_ld_data_only_new$PEAK<-as.factor(gwas_and_ld_data_only_new$PEAK)


suggestive_snps <- gwas_and_ld_data_only_new %>%
  filter(-log10(p_lrt) > suggestive)

index_snps_simp$PEAK<-as.factor(index_snps_simp$PEAK)

snps_count_per_peak <- suggestive_snps %>%
  inner_join(index_snps_simp, by = c("PEAK" = "PEAK")) %>%
  group_by(PEAK) %>%
  summarise(SNP_Count = n())

snps_count_per_peak$PEAK<-as.factor(snps_count_per_peak$PEAK)


genes_per_peak <- overlapping_annotations_df %>%
  left_join(gene_descriptions, by = c("gene_id" = "ensembl_gene_id")) %>%
  mutate(
    external_gene_name = if_else(is.na(external_gene_name), gene_id, external_gene_name),
    description.y = gsub("\\[.*?\\]", "", description.y),  # Remove text in brackets
    description.y = if_else(is.na(description.y) | description.y == "", "Unknown Gene", description.y)  # Replace NA or blank with "Unknown Gene"
  ) %>%
  group_by(PEAK) %>%
  summarise(Associated_Genes = toString(external_gene_name), Names = toString(description.y))


genes_per_peak$PEAK<-as.factor(genes_per_peak$PEAK)

# Create the final report
final_report <- index_snps_simp %>%
  dplyr::select(PEAK, CHR_A, maximum, minimum) %>%
  left_join(snps_count_per_peak, c("PEAK" = "PEAK")) %>%
  left_join(genes_per_peak, c("PEAK" = "PEAK"))

colnames(final_report) <- c("Peak", "Chromosome", "Start", "End", "n SNPs", "Associated Genes", "Descriptions")

# Replace commas with newlines for specific columns (e.g., "Associated Genes" and "Descriptions")
final_report <- final_report %>%
  mutate(
    `Associated Genes` = gsub(", ", "\n", `Associated Genes`),
    Descriptions = gsub(", ", "\n", Descriptions)
  )


# Replace commas with newlines for specific columns (e.g., "Associated Genes" and "Descriptions")
final_report <- final_report %>%
  rowwise() %>% # Process each row individually
  mutate(
    `Associated Genes` = strsplit(`Associated Genes`, ", "),  # Split into a list
    Descriptions = strsplit(Descriptions, ", ")              # Split into a list
  ) %>%
  unnest(cols = c(`Associated Genes`, Descriptions)) %>%     # Expand rows for each pair
  group_by(Peak, Chromosome, Start, End, `n SNPs`) %>%       # Regroup data
  summarize(
    `Associated Genes` = paste(`Associated Genes`, collapse = "\n"),
    Descriptions = paste(Descriptions, collapse = "\n"),
    .groups = "drop"
  )

# Create the flextable
flextable_table <- flextable(final_report) %>%
  bg(part = "header", bg = "#D3D3D3") %>%  # Set header background color
  bold(part = "header") %>%               # Bold the header row
  fontsize(size = 10, part = "all") %>%   # Set font size for the entire table
  autofit() %>%                           # Adjust column widths
  #width(j = 1:ncol(final_report), width = 1.5) %>%  # Set column widths
  set_table_properties(layout = "autofit") %>%      # Fit table to layout
  line_spacing(i = NULL, space = 1.2)     # Add line spacing for better readability




# Define landscape section properties
landscape_section <- block_section(
  prop_section(
    page_size = page_size(orient = "landscape") # Set the page orientation to landscape
  )
)

# Create the Word document
doc <- read_docx() %>%
  body_add_flextable(flextable_table) %>% # Add the flextable
  body_add(landscape_section)      # Apply landscape section properties


# Export the Word document
print(doc, target = file.path(root,"output_data/08_Peak_Analysis",params$data_path,paste0(params$data_path,".peak_report.docx")))

flextable_table

```