---
title: "VCF QC Report for SNPs"
output: html_notebook
params:
  data_path: ~/Desktop/MIC4273_HAP1_MSU618_all_fish_all_sites.FILTERED_BIALLELIC_SNPS_ONLY.gq.ss.txt
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(data.table)
require(ggplot2)
require(dplyr)
require(tidyr)
require(reshape2)
require(stringr)
setwd(root)
```

```{r}
#proposed cutoffs
AC_THRESH=0
N_MISSING_THRESH=0.1
```

```{r}
filtered_gt_datafile<-file.path(root,"output_data/03_QC","MIC4273_HAP2_NEWREF_MSU618_all_fish_all_sites.FILTERED_BIALLELIC_SNPS_ONLY.renamed.filtered.stats.sampled.tab")
missing
gt_data<-fread(filtered_gt_datafile)
missing_data_file<-file.path(root,"output_data/03_QC/","MIC4273_HAP2_NEWREF_MSU618_all_fish_all_sites.missingness_per_site_per_group.txt")
missing_data<-fread(missing_data_file)
```


```{r}
randomRows = function(df,n){
   return(df[sample(nrow(df),n),])
}
hist_vals<-as.data.frame(gt_data[,c(1:5)])
```

```{r}

# Step 2: Reshape the data to long format (to work per individual per variant)
missingness_long <- melt(missing_data, id.vars = c("CHROM", "POS"), variable.name = "Group", value.name = "Missingness")

```

```{r, fig.width=12}

# Plot using ggplot2 with geom_density, faceted by population
missingness_long %>%
  filter(grepl("_", Group)) %>%
  ggplot( aes(x = Missingness, fill = Group)) +
  geom_density(alpha = 0.3) +  # Use density plot to visualize the distribution
  geom_vline(xintercept = N_MISSING_THRESH, linetype="dotted",
                color = "red", size=1.5) +
  labs(title = "Distribution of Missing Genotypes per Variant by Population",
       x = "Percent of Missing Individuals per Variant",
       y = "Density") +
  facet_wrap(~ Group, scales = "free_y") +  # Facet by population, allowing y-axis to be independent
  theme_minimal() +                              # Use a clean theme
  theme(legend.position = "none")                # Remove legend since faceting already splits by population
```

```{r, fig.width=12}
# Plot using ggplot2 with geom_density, faceted by population
missingness_long %>%
  filter(!grepl("_", Group)) %>%
  ggplot( aes(x = Missingness, fill = Group)) +
  geom_histogram(binwidth=0.1) +  # Use density plot to visualize the distribution
  geom_vline(xintercept = N_MISSING_THRESH, linetype="dotted",
                color = "red", size=1.5) +
  labs(title = "Distribution of Missing Genotypes per Variant by Population",
       x = "Percent of Missing Individuals per Variant",
       y = "Density") +
  facet_wrap(~ Group, scales = "free_y") +  # Facet by population, allowing y-axis to be independent
  theme_minimal() +                              # Use a clean theme
  theme(legend.position = "none")                # Remove legend since faceting already splits by population
```


```{r}
count_equal_ac_an <- nrow(subset(hist_vals, AC == AN))
count_equal_ac_an
```

```{r, fig.width=8,fig.height=8}
ggplot(hist_vals, aes(x = AC)) +
    geom_density(color="darkblue",fill="lightblue",adjust=5,aes(y=..scaled..)) +
    geom_vline(xintercept = AC_THRESH, linetype="dotted",
                color = "red", size=1.5) +
    xlim(-10,200) +
    theme(text = element_text(size=15),
        axis.text.x = element_text(angle=90, hjust=1))
```

```{r}
count_ac_less_than_zero <- nrow(subset(hist_vals, AC == 0))
count_ac_less_than_zero
```


```{r}
cutoff_table<-NULL
cutoff_table<-as.data.frame(hist_vals %>% filter(value > GQ_thresh) %>% count(key))
cutoff_table$Total<-as.data.frame(hist_vals %>% count(key))$n
cutoff_table$POP<-str_split_fixed(cutoff_table$key,"_",2)[,1]
names(cutoff_table)<-c("Sample","GQ10","Total","POP")
cutoff_table$Percent<-cutoff_table$GQ10/cutoff_table$Total
ggplot(data=cutoff_table, aes(x=POP,y=Percent))+ geom_boxplot()

```

```{r}
confidence=10^(-GQ_thresh/10)
(1-confidence)*100
```
