---
title: "Peak Report"
output: html_notebook
params:
  donor_data_path: PKINGS_ALL_WOB_EXCLUDED
  recipient_data_path: PKINGS_ALL_WOB_INCLUDED
  output_file: NULL
---


```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
require(tidyverse)
library(data.table)
require(GenomicRanges)
library(stringr)
source(file.path(root,"code","14_Wobble_Analysis","read_bimbam_extract.R"))
```

## Data Loading

### Load the Peak Information from "donor" GWAS
```{r}
peaks_path<-file.path(root,"output_data","06_Association",params$donor_data_path,paste0(params$donor_data_path,"_PEAKS.bed"))

df.bed <- read.table(peaks_path)
names(df.bed) <- c("chr", "start","end", "peak","maxp","maxp_pos")
df.bed$chr<-paste0("chr",df.bed$chr)

#remove positions if there is only one SNP
df.bed <- df.bed %>% 
  mutate(lg = end- start) %>% 
  filter(lg > 0)

```

```{r}
res <- read_bimbam_extract(
  geno_file  = file.path(root,"output_data","06_Association",params$recipient_data_path,paste0(params$recipient_data_path,".geno")),
  map_file   = file.path(root,"output_data","06_Association",params$recipient_data_path,paste0(params$recipient_data_path,"_merge_map.txt")),
  order_file = file.path(root,"output_data","06_Association",params$recipient_data_path,paste0(params$recipient_data_path,".order")),
  positions  = df.bed$maxp_pos
)

# A tidy data.frame with variant metadata + genotypes:
res$table[1:3, 1:12]

# Just the dosage matrix [variants x samples], columns in .order order:
G <- res$genotypes

# If you want to append chr/pos back to rownames:
rownames(G) <- paste0(res$meta$chr, ":", res$meta$pos)

# Example: pull the genotypes for a single locus (e.g., pos 237236 on chr16)
idx <- which(res$meta$chr == "chr16" & res$meta$pos == 237236L)
G[idx, ]

```

```{r}
final_long <- make_dosage_long(res, df.bed)
head(final_long)
```
```{r}
covar_wide_by_peak <- final_long %>%
  select(IID, peak, dosage) %>%
  distinct() %>%
  mutate(peak = paste0("PEAK_", peak)) %>%
  pivot_wider(names_from = peak, values_from = dosage) %>%
  arrange(IID)

head(covar_wide_by_peak)
```


```{r}
iid_order <- fread(file.path(root,"output_data","06_Association",params$recipient_data_path,paste0(params$recipient_data_path,".order")), header=FALSE)[, trimws(V1)]

covar_wide_by_peak <- covar_wide_by_peak %>%
  mutate(IID = factor(IID, levels = iid_order)) %>%
  arrange(IID) 

covars <- covar_wide_by_peak %>%
  select(-IID)


write.table(covars, file.path(root,"input_data/14_Wobble_Analysis",params$recipient_data_path,paste0(params$recipient_data_path,".conditional.covars")),
            col.names = FALSE, row.names = FALSE, quote = FALSE)

head(covar_wide_by_peak)
```

## CC Phenotype

```{r}
pheno_path<-file.path(root,"input_data","01_Terra/data_model","fish_data_2024a.txt")
df.fish<-read.table(pheno_path,,header=T,sep="\t")
df.cc.pheno <- df.fish %>% filter(Species=="Paramormyrops kingsleyae") %>% select(PN, Phenotype)

df.cc.pheno$Phenotype<-as.factor(df.cc.pheno$Phenotype)

#TP and BP Grouped
#levels(df.cc.pheno$Phenotype) should be BP, TP, Wobble for this to work:

levels(df.cc.pheno$Phenotype)<-c(0,1,1)

df.cc.pheno <- df.cc.pheno %>%
  mutate(PN = factor(PN, levels = iid_order)) %>%
  arrange(PN)


write.table(df.cc.pheno$Phenotype,
            file = file.path(root,"input_data/14_Wobble_Analysis",params$recipient_data_path,paste0(params$recipient_data_path,".cc.pheno")),
            quote = FALSE, row.names = FALSE, col.names = FALSE)
```

## Quantitative Phenotype
```{r}
pheno_path<-file.path(root,"output_data","02_Phenotyping","fish_data_2024a.txt")

df.eods<-read.table(pheno_path,header=T,sep=",")

```


```{r}
    ggplot(data = df.eods, aes(x = aP0))+
     geom_histogram(binwidth=0.005)+ # or bins = 30
     xlim(-1,.1)
```

```{r}
r <- rank(df.eods$aP0, ties.method = "average")
z <- qnorm((r - 0.5) / (length(r) + 1))

df.eods$Z <- z

pheno_aligned <- ord %>%
  left_join(df.eods %>% select(sample_name, Z), by = c("IID" = "sample_name"))


ggplot(data = df.eods, aes(x = Z))+
     geom_histogram(bins=30) # or bins = 30

write.table(pheno_aligned$Z,
            file = file.path(root,"input_data/14_Wobble_Analysis",params$data_path,paste0(params$data_path,".int.pheno")),
            quote = FALSE, row.names = FALSE, col.names = FALSE)

```

