---
title: "RNASeq Report"
output: html_notebook
params:
  g1_data_path: APA_BEN_BAM_TP1_BP2_WOB9
  g2_data_path: MOV_BIR_DOG_DOV_BAVA_BIK_APA_BEN_TP1_BP2_WOB9
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
library(tximport)
library(tidyverse)
library("DESeq2")
library(stringr)
library(apeglm)
library(EnhancedVolcano)
require(rtracklayer)
library(ggrepel)
library(patchwork)   # For combining plots
library(biomaRt)
library(topGO)
```

```{r}
sample_data<-file.path(root, "input_data/09_RNASeq/rnaseq_metadata.txt")
coldata <- read.table(sample_data,header = T)


pheno_data<-coldata %>% select(c(-R1,-R2)) %>% unique()

#values_to_drop <- c("PK1", "PK2", "PK3")

#pheno_data<- pheno_data  %>%
  #filter(!SpecNo %in% values_to_drop)

pheno_data$Population <- factor(pheno_data$Population)
pheno_data$Phenotype <- factor(pheno_data$Phenotype)
```

```{r}
gwas_1_set<-c("APA","BAM","BIK")

gwas_2_set<-c("MOV","BIR","DOG","DOV","BAVA","BIK","APA","BEN")

gw1_pheno_data <- pheno_data[pheno_data$Population %in% gwas_1_set, ]
gw2_pheno_data <- pheno_data[pheno_data$Population %in% gwas_2_set, ]

gw1_genes_files <- file.path(root, "output_data/09_RNASeq/quantification",  paste0(gw1_pheno_data$SpecNo, ".genes.results"))
names(gw1_genes_files)<- paste0(gw1_pheno_data$Population,"_",gw1_pheno_data$SpecNo)

gw2_genes_files <- file.path(root, "output_data/09_RNASeq/quantification",  paste0(gw2_pheno_data$SpecNo, ".genes.results"))
names(gw2_genes_files)<- paste0(gw2_pheno_data$Population,"_",gw2_pheno_data$SpecNo)
```

```{r}
gw1_genes.rsem <- tximport(gw1_genes_files, type = "none", txIn = FALSE, txOut = FALSE,geneIdCol="gene_id",abundanceCol="TPM",countsCol="expected_count",lengthCol="effective_length",importer=function(x) readr::read_tsv(x))

gw2_genes.rsem <- tximport(gw2_genes_files, type = "none", txIn = FALSE, txOut = FALSE,geneIdCol="gene_id",abundanceCol="TPM",countsCol="expected_count",lengthCol="effective_length",importer=function(x) readr::read_tsv(x))

```


```{r}
# fix "Error: all(lengths > 0) is not TRUE" error
gw1_genes.rsem$length[gw1_genes.rsem$length == 0] = 0.01
gw2_genes.rsem$length[gw2_genes.rsem$length == 0] = 0.01

gw1_dds <- DESeqDataSetFromTximport(txi = gw1_genes.rsem,
                              colData = gw1_pheno_data,
                              design = ~  Phenotype )

gw2_dds <- DESeqDataSetFromTximport(txi = gw2_genes.rsem,
                              colData = gw2_pheno_data,
                              design = ~  Phenotype )

```

```{r}
keep<-rowSums(counts(gw1_dds)) >=10
gw1_dds<-gw1_dds[keep,]
gw1_dds <- DESeq(gw1_dds)

keep<-rowSums(counts(gw2_dds)) >=10
gw2_dds<-gw2_dds[keep,]
gw2_dds <- DESeq(gw2_dds)
```

```{r}
gw1_res <- results(gw1_dds, name = "Phenotype_TP_vs_BP", alpha = 0.1)
gw2_res <- results(gw2_dds, name = "Phenotype_TP_vs_BP", alpha = 0.1)

summary(gw1_res)
summary(gw2_res)
```
```{r}
peak_data_path<-file.path(root,"output_data","06_Association",params$g1_data_path)
g1_overlapping_annotations_df<-read.table(file.path(peak_data_path,paste0(params$g1_data_path,"_GENES_IN_PEAKS.txt")),header=T)


peak_data_path<-file.path(root,"output_data","06_Association",params$g2_data_path)
g2_overlapping_annotations_df<-read.table(file.path(peak_data_path,paste0(params$g2_data_path,"_GENES_IN_PEAKS.txt")),header=T)

```

```{r}
pCutoff = .1
FCcutoff = 1.0

# Filter the results
gw1_sig <- gw1_res[!is.na(gw1_res$padj) & !is.na(gw1_res$log2FoldChange) & 
                   gw1_res$padj <= pCutoff & gw1_res$log2FoldChange >= FCcutoff, ]

# Filter gw1_sig based on IDs in overlapping_annotations_df
filtered_gw1_sig <- gw1_sig[gw1_sig$ID %in% g1_overlapping_annotations_df$ID, ]

# Filter the results
gw2_sig <- gw2_res[!is.na(gw2_res$padj) & !is.na(gw2_res$log2FoldChange) & 
                   gw2_res$padj <= pCutoff & gw2_res$log2FoldChange >= FCcutoff, ]

# Filter gw1_sig based on IDs in overlapping_annotations_df
filtered_gw2_sig <- gw2_sig[gw2_sig$ID %in% g2_overlapping_annotations_df$ID, ]

```


```{r}
gw1_res4plot<-as.data.frame(gw1_res)
gw1_res4plot$gene_id <- rownames(gw1_res4plot)
gw1_res4plot <- merge(gw1_res4plot, g1_overlapping_annotations_df[, c("gene_id", "external_gene_name")],
                 by.x = "gene_id", by.y = "gene_id", all.x = TRUE)
gw1_res4plot$color_group <- ifelse(gw1_res4plot$gene_id %in% g1_overlapping_annotations_df$gene_id, "highlight", "normal")

g1_plot<-ggplot(data = gw1_res4plot) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') +
  # First layer: black points (normal)
  geom_point(data = subset(gw1_res4plot, color_group == "normal"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "grey", size = 2, alpha = 0.5) +
  # Second layer: red points (highlight)
  geom_point(data = subset(gw1_res4plot, color_group == "highlight"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "red", size = 2, alpha = 0.4) +
  # Add labels for red points using geom_text_repel
 geom_text_repel(data = subset(gw1_res4plot, color_group == "highlight"),
                  aes(x = log2FoldChange, y = -log10(padj), label = external_gene_name),
                  size = 3, 
                  min.segment.length = 0, 
                  max.overlaps = Inf, 
                  box.padding = 0.5, 
                  segment.size = 0.2, 
                  force = 2,  # Increase repulsion force
                  force_pull = 1,  # Fine-tune pull-back force
                  point.padding = 0.5,  # Space labels from points
                  max.time = 2)+  # Allow more time for resolving overlaps
                  #direction = "y") +  # Allow vertical adjustment for better spacing
  coord_cartesian(ylim = c(0, 15), xlim = c(-8, 8)) +
  labs(color = 'Gene Highlight',
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) +
  scale_x_continuous(breaks = seq(-10, 10, 2)) +
  ggtitle('Differential expression between Apassa Creek and Bambomo Creek')

gw2_res4plot<-as.data.frame(gw2_res)
gw2_res4plot$gene_id <- rownames(gw2_res4plot)
gw2_res4plot <- merge(gw2_res4plot, g2_overlapping_annotations_df[, c("gene_id", "external_gene_name")],
                 by.x = "gene_id", by.y = "gene_id", all.x = TRUE)
gw2_res4plot$color_group <- ifelse(gw2_res4plot$gene_id %in% g2_overlapping_annotations_df$gene_id, "highlight", "normal")

g2_plot<-ggplot(data = gw2_res4plot) +
  geom_vline(xintercept = c(-0.6, 0.6), col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = 'dashed') +
  # First layer: black points (normal)
  geom_point(data = subset(gw2_res4plot, color_group == "normal"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "grey", size = 2, alpha = 0.5) +
  # Second layer: red points (highlight)
  geom_point(data = subset(gw2_res4plot, color_group == "highlight"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "red", size = 2, alpha = 0.4) +
  # Add labels for red points using geom_text_repel
 geom_text_repel(data = subset(gw2_res4plot, color_group == "highlight" & padj < 0.1 ),
                  aes(x = log2FoldChange, y = -log10(padj), label = external_gene_name),
                  size = 3,
                  min.segment.length = 0,
                  max.overlaps = Inf,
                  box.padding = 0.5,
                  #segment.size = 0.2,
                  force = 2,  # Increase repulsion force
                  force_pull = 1,  # Fine-tune pull-back force
                  #point.padding = 0.5,  # Space labels from points
                  max.time = 2)+  # Allow more time for resolving overlaps
                  #direction = "y") +  # Allow vertical adjustment for better spacing
  coord_cartesian(ylim = c(0, 15), xlim = c(-8, 8)) +
  labs(color = 'Gene Highlight',
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) +
  scale_x_continuous(breaks = seq(-10, 10, 2)) +
  ggtitle('Differential expression between GWAS2')
```

```{r, fig.width=8}

combined_plot <- g1_plot + g2_plot

combined_plot <- combined_plot + 
                 plot_annotation(tag_levels = 'A') &
                 theme(plot.tag = element_text(size = 24, face="bold"))

# Adjust relative heights of the plots to change the overall aspect ratio
# Increase the values to make the plots taller
combined_plot <- combined_plot + 
                 plot_layout(ncol=2) 

# Print the combined plot
combined_plot
```


```{r, fig.height=8}
# Connect to Ensembl release 112 for Paramormyrops kingsleyae
ensembl_112 <- useEnsembl(
    biomart = "ensembl", 
    version = 112, 
    host = "https://may2024.archive.ensembl.org"
)

# Select the Paramormyrops kingsleyae dataset
ensembl_112 <- useDataset("pkingsleyae_gene_ensembl", mart = ensembl_112)

# Retrieve ENSEMBL IDs and GO terms
go_annotations <- getBM(
    attributes = c("ensembl_gene_id", "go_id"),
    mart = ensembl_112
)

# Remove entries with missing GO IDs
go_annotations <- go_annotations[go_annotations$go_id != "", ]

# Create a named list mapping genes to GO terms
gene_to_go <- split(go_annotations$go_id, go_annotations$ensembl_gene_id)

# Example list of DE ENSEMBL IDs
de_genes <- rownames(gw1_res %>% subset(abs(log2FoldChange) > 1 & padj <0.1))

# Create a named vector: 1 for DE genes, 0 for non-DE genes
gene_list <- factor(as.integer(names(gene_to_go) %in% de_genes))
names(gene_list) <- names(gene_to_go)

# Create a topGOdata object
go_data <- new(
    "topGOdata",
    ontology = "BP",                     # Ontology type: "BP", "MF", or "CC"
    allGenes = gene_list,                # Named vector of genes
    geneSelectionFun = function(x) x == 1,  # Function to select DE genes
    annot = annFUN.gene2GO,              # Use gene-to-GO mapping
    gene2GO = gene_to_go
)

# Run the analysis
go_results <- runTest(go_data, algorithm = "classic", statistic = "fisher")

# Retrieve all results
all_results <- GenTable(
    go_data,
    classicFisher = go_results,
    orderBy = "classicFisher",
    ranksOf = "classicFisher",
    topNodes = length(score(go_results))  # Include all nodes
)

# Convert p-values to numeric
all_results$classicFisher <- as.numeric(all_results$classicFisher)

# Filter for p-value < 0.01
filtered_results <- all_results[all_results$classicFisher < 0.001, ]

# Display the filtered results
print(filtered_results)


```
```{r, fig.height=8}
ggplot(filtered_results, aes(x = -log10(classicFisher), y = reorder(Term, -classicFisher), size = Significant, color = -log10(classicFisher))) +
    geom_point() +
    scale_size_binned(name = "Gene Count",n.breaks=8) +
    labs(
        title = "GO Enrichment Dot Plot",
        x = "Degree of Enrichment (-log10(P-value))",
        y = "GO Terms"
    ) +
    theme_minimal() +
    theme(
        axis.text.y = element_text(size = 10),
        plot.title = element_text(hjust = 0.5, size = 14)
    )
```

```{r}
plotCounts(gw_dds,"LOC111845409",intgroup="Phenotype")
```



```{r}
rld<-rlog(dds)
rld_mat<-assay(rld)
pca<-prcomp(t(rld_mat))
df<-cbind(pheno_data,pca$x)

ggplot(df, aes(x=PC1, y=PC2, color = Population, label=rownames(df))) + geom_text()
```
```{r, fig.width=12}
plotCounts(dds,"gene-sptbn4.4",intgroup=c("Phenotype","Population"), normalized=T, transform=F,returnData = F)
```

```{r}
summary(res)
```