---
title: "RNASeq Report"
output: html_notebook
params:
  g1_data_path: APA_BEN_BAM_TP1_BP2_WOB9
  g2_data_path: MOV_BIR_DOG_DOV_BAVA_BIK_APA_BEN_TP1_BP2_WOB9
---

```{r setup, include=FALSE}
root<-rprojroot::find_root(".git/index")
knitr::opts_knit$set(root.dir = root)
library(tximport)
library(tidyverse)
library("DESeq2")
library(stringr)
require(rtracklayer)
library(ggrepel)
library(patchwork)   # For combining plots
library(biomaRt)
library(ggpubr)
source(file.path(root,"code","09_RNASeq","go_analysis.R"))
```

```{r}
prev_rnaseq_results<-file.path(root, "input_data/09_RNASeq/12862_2019_1572_MOESM6_ESM.csv")
losilla_complexity_list<-read.csv(prev_rnaseq_results,header=T)

losilla_complexity_list<-losilla_complexity_list %>% filter(type_of_gene=="protein-coding")



ensembl_112 <- useEnsembl(biomart = "ensembl", version = 112,host="https://may2024.archive.ensembl.org")
ensembl_112 <-useDataset("pkingsleyae_gene_ensembl",mart=ensembl_112)


available_entrez_ids <- getBM(attributes = "entrezgene_id", mart = ensembl_112)
unmatched_ids <- setdiff(losilla_complexity_list$Pking_Entrez_geneID, available_entrez_ids$entrezgene_id)
print(unmatched_ids)


# Assuming the attribute 'description' exists, retrieve the gene descriptions
losilla_complexity_list_converted <- getBM(values=losilla_complexity_list$Pking_Entrez_geneID,
                                filters = "entrezgene_id",
  attributes = c("entrezgene_id","ensembl_gene_id","external_gene_name", "description"),
  mart = ensembl_112
)  
```

```{r}
sample_data<-file.path(root, "input_data/09_RNASeq/rnaseq_metadata.txt")
coldata <- read.table(sample_data,header = T)


pheno_data<-coldata %>% dplyr::select(c(-R1,-R2)) %>% unique()

values_to_drop <- c("PK1", "PK2", "PK3")

pheno_data<- pheno_data  %>%
  filter(!SpecNo %in% values_to_drop)

pheno_data$Population <- factor(pheno_data$Population)
pheno_data$Phenotype <- factor(pheno_data$Phenotype)
```

```{r}
gwas_1_set<-c("APA","BAM","BEN")

gwas_2_set<-c("MOV","BIR","DOG","DOV","BAVA","BIK","APA","BEN")

gw1_pheno_data <- pheno_data[pheno_data$Population %in% gwas_1_set, ]
gw2_pheno_data <- pheno_data[pheno_data$Population %in% gwas_2_set, ]

gw1_genes_files <- file.path(root, "output_data/09_RNASeq/quantification",  paste0(gw1_pheno_data$SpecNo, ".genes.results"))
names(gw1_genes_files)<- paste0(gw1_pheno_data$Population,"_",gw1_pheno_data$SpecNo)

gw2_genes_files <- file.path(root, "output_data/09_RNASeq/quantification",  paste0(gw2_pheno_data$SpecNo, ".genes.results"))
names(gw2_genes_files)<- paste0(gw2_pheno_data$Population,"_",gw2_pheno_data$SpecNo)
```

```{r}
gw1_genes.rsem <- tximport(gw1_genes_files, type = "none", txIn = FALSE, txOut = FALSE,geneIdCol="gene_id",abundanceCol="TPM",countsCol="expected_count",lengthCol="effective_length",importer=function(x) readr::read_tsv(x))

gw2_genes.rsem <- tximport(gw2_genes_files, type = "none", txIn = FALSE, txOut = FALSE,geneIdCol="gene_id",abundanceCol="TPM",countsCol="expected_count",lengthCol="effective_length",importer=function(x) readr::read_tsv(x))

```


```{r}
# fix "Error: all(lengths > 0) is not TRUE" error
gw1_genes.rsem$length[gw1_genes.rsem$length == 0] = 0.01
gw2_genes.rsem$length[gw2_genes.rsem$length == 0] = 0.01

gw1_dds <- DESeqDataSetFromTximport(txi = gw1_genes.rsem,
                              colData = gw1_pheno_data,
                              design = ~  Phenotype )

gw2_dds <- DESeqDataSetFromTximport(txi = gw2_genes.rsem,
                              colData = gw2_pheno_data,
                              design = ~  Phenotype )

```

```{r}
keep<-rowSums(counts(gw1_dds)) >=10
gw1_dds<-gw1_dds[keep,]
gw1_dds <- DESeq(gw1_dds)

keep<-rowSums(counts(gw2_dds)) >=10
gw2_dds<-gw2_dds[keep,]
gw2_dds <- DESeq(gw2_dds)
```

```{r}
#A positive log2 fold change for a comparison of TP vs BP means that gene expression in TP is larger in comparison to BP.

gw1_res_o <- results(gw1_dds, name = "Phenotype_TP_vs_BP", alpha = 0.1)
gw2_res_o <- results(gw2_dds, name = "Phenotype_TP_vs_BP", alpha = 0.1)

gw1_res = lfcShrink(dds=gw1_dds, coef="Phenotype_TP_vs_BP", type = "ashr")
gw2_res = lfcShrink(dds=gw2_dds, coef="Phenotype_TP_vs_BP", type = "ashr")

summary(gw1_res)
summary(gw2_res)
```
```{r}
peak_data_path<-file.path(root,"output_data","06_Association",params$g1_data_path)
g1_overlapping_annotations_df<-read.table(file.path(peak_data_path,paste0(params$g1_data_path,"_GENES_IN_PEAKS.txt")),header=T)


peak_data_path<-file.path(root,"output_data","06_Association",params$g2_data_path)
g2_overlapping_annotations_df<-read.table(file.path(peak_data_path,paste0(params$g2_data_path,"_GENES_IN_PEAKS.txt")),header=T)

```

```{r}
pCutoff = .1
FCcutoff = 0.5

# Filter the results
gw1_sig <- gw1_res[!is.na(gw1_res$padj) & !is.na(gw1_res$log2FoldChange) & 
                   gw1_res$padj <= pCutoff & abs(gw1_res$log2FoldChange) >= FCcutoff, ]

# Filter gw1_sig based on IDs in overlapping_annotations_df
filtered_gw1_sig <- gw1_sig[rownames(gw1_sig) %in% g1_overlapping_annotations_df$gene_id, ]

# Filter the results
gw2_sig <- gw2_res[!is.na(gw2_res$padj) & !is.na(gw2_res$log2FoldChange) & 
                   gw2_res$padj <= pCutoff & abs(gw2_res$log2FoldChange) >= FCcutoff, ]

# Filter gw1_sig based on IDs in overlapping_annotations_df
filtered_gw2_sig <- gw2_sig[rownames(gw2_sig) %in% g2_overlapping_annotations_df$gene_id, ]

```

```{r}
ensembl_112 <- useEnsembl(biomart = "ensembl", version = 112,host="https://may2024.archive.ensembl.org")
ensembl_112 <-useDataset("pkingsleyae_gene_ensembl",mart=ensembl_112)

# Assuming the attribute 'description' exists, retrieve the gene descriptions
gene_descriptions <- getBM(values=rownames(gw1_res),
                                filters = "ensembl_gene_id",
  attributes = c("ensembl_gene_id", "description","external_gene_name"),
  mart = ensembl_112
)

gw1_res$ensembl_gene_id=rownames(gw1_res)
gw1_res<-as.data.frame(gw1_res)
gw1_res<- gw1_res %>% left_join(gene_descriptions,by="ensembl_gene_id")
gw1_res<-gw1_res[order(gw1_res$log2FoldChange, decreasing = TRUE), ]

# Assuming the attribute 'description' exists, retrieve the gene descriptions
gene_descriptions <- getBM(values=rownames(gw2_res),
                                filters = "ensembl_gene_id",
  attributes = c("ensembl_gene_id", "description","external_gene_name"),
  mart = ensembl_112
)

gw2_res$ensembl_gene_id=rownames(gw2_res)
gw2_res<-as.data.frame(gw2_res)
gw2_res<- gw2_res %>% left_join(gene_descriptions,by="ensembl_gene_id")
gw2_res<-gw2_res[order(gw2_res$log2FoldChange, decreasing = TRUE), ]
```


```{r}
gw1_res4plot<-as.data.frame(gw1_res)
gw1_res4plot <- gw1_res4plot %>%
  mutate(label_name = ifelse(is.na(external_gene_name), ensembl_gene_id, external_gene_name))

gw1_res4plot$color_group <- case_when(
  gw1_res4plot$ensembl_gene_id %in% g1_overlapping_annotations_df$gene_id ~ "highlight",
  #gw1_res4plot$ensembl_gene_id %in% losilla_complexity_list_converted$ensembl_gene_id ~ "previous",
  TRUE ~ "normal"
)

gw1_res4plot[gw1_res4plot$ensembl_gene_id=="ENSPKIG00000008497",]$label_name<-"cntnap5b"

g1_plot <- ggplot(data = gw1_res4plot) +
  geom_vline(xintercept = -0.5, col = "gray", linetype = 'dashed') +
    geom_vline(xintercept = .5, col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = 0.1, col = "gray", linetype = 'dashed') +
  # First layer: black points (normal)
  geom_point(data = subset(gw1_res4plot, color_group == "normal"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "grey", size = 2, alpha = 0.5) +
  # Second layer: red points (highlight)
    geom_point(data = subset(gw1_res4plot, log2FoldChange<(-0.5) & -log10(padj)>0.1), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "blue", size = 2, alpha = 0.4) +
  geom_point(data = subset(gw1_res4plot, log2FoldChange>0.5 & -log10(padj)>0.1), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "red", size = 2, alpha = 0.4) +
  #Add labels for red points using geom_text_repel
 geom_text_repel(data = subset(gw1_res4plot, color_group == "highlight" & -log10(padj)>0.1 & abs(log2FoldChange)>0.5),
                  aes(x = log2FoldChange, y = -log10(padj), label = label_name),
                 size = 4,
                  min.segment.length = 0,
                  max.overlaps = Inf,
                  box.padding = 0.5,
                  segment.size = 0.2,
                  force = 2,  # Increase repulsion force
                  force_pull = 1,  # Fine-tune pull-back force
                  max.time = 2)+  # Allow more time for resolving overlaps
  coord_cartesian(ylim = c(0, 15), xlim = c(-6, 6)) +
  labs(color = 'Gene Highlight',
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) +
  scale_x_continuous(breaks = seq(-6, 6, 2)) +
  # Add annotations
  annotate("text", x = -6, y = 14, label = "Higher in BP", hjust = 0, size = 4, color = "black") +
  annotate("text", x = 6, y = 14, label = "Higher in TP", hjust = 1, size = 4, color = "black") +
  #ggtitle('Apassa, Bengue, and Bambomo Creeks') +
  # Custom theme for title size, removing background and grid lines
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5), # Control title size and alignment
    panel.background = element_blank(), # Remove background
    panel.grid.major = element_blank(), # Remove major grid lines
    panel.grid.minor = element_blank(), # Remove minor grid lines
    axis.line = element_line(color = "black"), # Add axis lines
    axis.text = element_text(size = 12), # Customize axis text size
    axis.title = element_text(size = 14) # Customize axis title size
  )

gw2_res4plot<-as.data.frame(gw2_res)

gw2_res4plot <- gw2_res4plot %>%
  mutate(label_name = ifelse(is.na(external_gene_name), ensembl_gene_id, external_gene_name))

gw2_res4plot$color_group <- ifelse(gw2_res4plot$ensembl_gene_id %in% g2_overlapping_annotations_df$gene_id, "highlight", "normal")

g2_plot<-ggplot(data = gw2_res4plot) +
  geom_vline(xintercept = -.5, col = "gray", linetype = 'dashed') +
    geom_vline(xintercept = .5, col = "gray", linetype = 'dashed') +
  geom_hline(yintercept = 0.1, col = "gray", linetype = 'dashed') +
  # First layer: black points (normal)
  geom_point(data = subset(gw2_res4plot, color_group == "normal"), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "grey", size = 2, alpha = 0.5) +
  # Second layer: red points (highlight)
  geom_point(data = subset(gw2_res4plot, log2FoldChange<(-0.5) & -log10(padj)>0.1), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "blue", size = 2, alpha = 0.4) +
  geom_point(data = subset(gw2_res4plot, log2FoldChange>0.5 & -log10(padj)>0.1), 
             aes(x = log2FoldChange, y = -log10(padj)), 
             color = "red", size = 2, alpha = 0.4) +
  # Add labels for red points using geom_text_repel
 geom_text_repel(data = subset(gw2_res4plot, color_group == "highlight" & -log10(padj)>0.1 & abs(log2FoldChange)>0.5 ),
                  aes(x = log2FoldChange, y = -log10(padj), label = label_name),
                  size = 4,
                  min.segment.length = 0,
                  max.overlaps = Inf,
                  box.padding = 0.5,
                  segment.size = 0.2,
                  force = 2,  # Increase repulsion force
                  force_pull = 1,  # Fine-tune pull-back force
                  max.time = 2)+  # Allow more time for resolving overlaps
  coord_cartesian(ylim = c(0, 15), xlim = c(-6, 6)) +
  labs(color = 'Gene Highlight',
       x = expression("log"[2]*"FC"), y = expression("-log"[10]*"p-value")) +
  scale_x_continuous(breaks = seq(-6, 6, 2)) +
  annotate("text", x = -6, y = 14, label = "Higher in BP", hjust = 0, size = 4, color = "black") +
  annotate("text", x = 6, y = 14, label = "Higher in TP", hjust = 1, size = 4, color = "black") +
  #ggtitle("Mouvanga, Douengi, Dovalou,Apassa, Bengue, Biroundou, Bavavela and Bikagala Creeks")+
  theme(
    plot.title = element_text(size = 12, face = "bold", hjust = 0.5), # Control title size and alignment
    panel.background = element_blank(), # Remove background
    panel.grid.major = element_blank(), # Remove major grid lines
    panel.grid.minor = element_blank(), # Remove minor grid lines
    axis.line = element_line(color = "black"), # Add axis lines
    axis.text = element_text(size = 12), # Customize axis text size
    axis.title = element_text(size = 14) # Customize axis title size
  )
```


```{r, fig.width=24}
genes_of_interest<-c("ENSPKIG00000008497","ENSPKIG00000008545","ENSPKIG00000004799","ENSPKIG00000004648","ENSPKIG00000000098","ENSPKIG00000009096","ENSPKIG00000006520","ENSPKIG00000025014","ENSPKIG00000004553","ENSPKIG00000005195","ENSPKIG00000000155","ENSPKIG00000006468")
names_of_interest<-c("cntnap5b","gypc","sptbn4b","itpr3","htt","itih3a.1","slc7a10b","slx4","frs3","mustn1a","rgs12b","slc7a9")

genes_of_interest<-c("ENSPKIG00000008497","ENSPKIG00000009096")
names_of_interest<-c("cntnap5b","itih3a.1")

gene_name_mapping <- setNames(names_of_interest, genes_of_interest)

gene_name_mapping<-gene_name_mapping %>% as.data.frame() %>% rownames_to_column(var="ensembl_gene_id")
colnames(gene_name_mapping)<-c("ensembl_gene_id","genename")

# Extract TPM data from tximport object (genes as rows, samples as columns)
#gw1_tpm_data <- as.data.frame(gw1_genes.rsem$abundance)
#gw2_tpm_data <- as.data.frame(gw2_genes.rsem$abundance)

gw1_normalized_counts<-counts(gw1_dds,normalized=TRUE)
gw2_normalized_counts<-counts(gw2_dds,normalized=TRUE)
# Extract TPM values for matching transcripts
gw1_tpm_filtered <- gw1_normalized_counts[rownames(gw1_normalized_counts) %in% genes_of_interest, , drop = FALSE]
gw2_tpm_filtered <- gw2_normalized_counts[rownames(gw2_normalized_counts) %in% genes_of_interest, , drop = FALSE]

# Add gene_id for annotation
gw1_tpm_filtered <- data.frame(ensembl_gene_id = rownames(gw1_tpm_filtered), gw1_tpm_filtered)
gw1_tpm_filtered <- left_join(gw1_tpm_filtered, gene_name_mapping, by = "ensembl_gene_id")

gw2_tpm_filtered <- data.frame(ensembl_gene_id = rownames(gw2_tpm_filtered), gw2_tpm_filtered)
gw2_tpm_filtered <- left_join(gw2_tpm_filtered, gene_name_mapping, by = "ensembl_gene_id")

pheno_data$Sample<-paste0(pheno_data$Population,"_",pheno_data$SpecNo)

# Pivot data longer: Each row represents a gene-sample combination
gw1_tpm_long <- gw1_tpm_filtered %>%
  pivot_longer(cols = c(-ensembl_gene_id,-genename) , names_to = "Sample", values_to = "Normalized Counts")

# Join the tissue information
gw1_tpm_long <- gw1_tpm_long %>%
  left_join(pheno_data, by = "Sample")


# Pivot data longer: Each row represents a gene-sample combination
gw2_tpm_long <- gw2_tpm_filtered %>%
  pivot_longer(cols = c(-ensembl_gene_id,-genename) , names_to = "Sample", values_to = "Normalized Counts")

# Join the tissue information
gw2_tpm_long <- gw2_tpm_long %>%
  left_join(pheno_data, by = "Sample")

tpm_long<-rbind(gw1_tpm_long,gw2_tpm_long)

# Plot the data
cntn<-ggplot(subset(tpm_long, ensembl_gene_id=="ENSPKIG00000008497"), aes(x = Population, y = `Normalized Counts`, fill=Phenotype)) +
  geom_boxplot() +
  facet_wrap(~ genename, scales = "free") +
  theme_minimal() +
theme(
    legend.position = "none",  # Hide legend if needed
    panel.grid.major.x = element_blank(),  # Remove major vertical grid lines
    panel.grid.minor.x = element_blank(),  # Remove minor vertical grid lines
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis text
    axis.title.y = element_text(margin = margin(r = 0)),
    strip.text = element_text(size = 14, face = "bold"),  # Increase facet label size and make bold
  ) +
  scale_fill_manual(values = c("BP" = "blue", "TP" = "red")) +  # Set manual colors
  #scale_y_log10() +
  labs(
       x = "Population",
       y = "Normalized Counts",
       fill = "Tissue") 

ith<-ggplot(subset(tpm_long, ensembl_gene_id=="ENSPKIG00000009096"), aes(x = Population, y = `Normalized Counts`, fill=Phenotype)) +
  geom_boxplot() +
  facet_wrap(~ genename, scales = "free") +
  theme_minimal() +
  theme(
    legend.position = "none",  # Hide legend if needed
    panel.grid.major.x = element_blank(),  # Remove major vertical grid lines
    panel.grid.minor.x = element_blank(),  # Remove minor vertical grid lines
    axis.text.x = element_text(angle = 45, hjust = 1),  # Rotate x-axis text
    strip.text = element_text(size = 14, face = "bold")  # Increase facet label size and make bold
  ) +
    scale_fill_manual(values = c("BP" = "blue", "TP" = "red")) +  # Set manual colors
    #scale_y_log10() +
  labs(
       x = "Population",
       y = "Normalized Counts",
       fill = "Tissue")
```

```{r}
# Connect to Ensembl release 112 for Paramormyrops kingsleyae
ensembl_112 <- useEnsembl(
    biomart = "ensembl", 
    version = 112, 
    host = "https://may2024.archive.ensembl.org"
)

# Select the Paramormyrops kingsleyae dataset
ensembl_112 <- useDataset("pkingsleyae_gene_ensembl", mart = ensembl_112)

# Retrieve ENSEMBL IDs and GO terms
go_annotations <- getBM(
    attributes = c("ensembl_gene_id", "go_id"),
    mart = ensembl_112
)

# Remove entries with missing GO IDs
go_annotations <- go_annotations[go_annotations$go_id != "", ]

# Create a named list mapping genes to GO terms
gene_to_go <- split(go_annotations$go_id, go_annotations$ensembl_gene_id)
```

```{r}
# Define gene sets based on log2 fold change.
up_genes <- gw1_sig$ensembl_gene_id[gw1_sig$log2FoldChange > 0]
down_genes <- gw1_sig$ensembl_gene_id[gw1_sig$log2FoldChange < 0]

# Run GO analysis for upregulated genes.
filtered_results_up <- run_go_analysis(
  de_genes = up_genes,
  gene_to_go = gene_to_go,
  ontology = "CC",       # Change to "BP" or "MF" if desired.
  p_threshold = 0.05
)

# Run GO analysis for downregulated genes.
filtered_results_down <- run_go_analysis(
  de_genes = down_genes,
  gene_to_go = gene_to_go,
  ontology = "CC",
  p_threshold = 0.05
)

gw1_go<-plot_combined_go_results(subset(filtered_results_up, Significant > 2), subset(filtered_results_down, Significant > 2))

gw1_go
```
```{r}
# Define gene sets based on log2 fold change.
up_genes <- gw2_sig$ensembl_gene_id[gw2_sig$log2FoldChange > 0]
down_genes <- gw2_sig$ensembl_gene_id[gw2_sig$log2FoldChange < 0]

# Run GO analysis for upregulated genes.
filtered_results_up <- run_go_analysis(
  de_genes = up_genes,
  gene_to_go = gene_to_go,
  ontology = "CC",       # Change to "BP" or "MF" if desired.
  p_threshold = 0.05
)

# Run GO analysis for downregulated genes.
filtered_results_down <- run_go_analysis(
  de_genes = down_genes,
  gene_to_go = gene_to_go,
  ontology = "CC",
  p_threshold = 0.05
)

gw2_go<-plot_combined_go_results(subset(filtered_results_up, Significant > 2), subset(filtered_results_down, Significant > 2))

gw2_go
```

```{r, fig.width=24, fig.height=8}

p1 <- (g1_plot / cntn + plot_layout(heights = c(2, 1))) | gw1_go + 
      plot_layout(widths = c(2, 1))

p2 <- (g2_plot / ith + plot_layout(heights = c(2, 1))) | gw2_go + 
      plot_layout(widths = c(2, 1))

final_plot <- wrap_elements(p1) + wrap_elements(p2) + plot_layout(ncol = 2)

final_plot

ggsave(
    file.path(root, "output_data", "09_RNASeq", "volcano_plot_RNAseq.svg"),
    plot = final_plot, width = 24, height = 8
 )

```